// @babel/plugin-bugfix-safari-rest-destructuring-rhs-array/lib/index.js
import { declare } from '@babel/helper-plugin-utils';
import { skipTransparentExprWrappers } from '@babel/helper-skip-transparent-expression-wrappers';

function isPathSimpleArrayWithLength(path, N) {
  if (!path.isArrayExpression()) {
    return false;
  }
  const elementPaths = path.get("elements");
  return elementPaths.length === N && elementPaths.every(elementPath => elementPath.node && !elementPath.isSpreadElement()) && !path.node.extra?.trailingComma;
}
function shouldTransform(path) {
  const elementPaths = path.get("elements");
  const elementPathsLength = elementPaths.length;
  if (elementPaths[elementPathsLength - 1].isRestElement()) {
    const {
      parentPath
    } = path;
    const rhsPath = parentPath.isVariableDeclarator() ? parentPath.get("init") : parentPath.isAssignmentExpression() ? parentPath.get("right") : null;
    if (rhsPath?.node && isPathSimpleArrayWithLength(skipTransparentExprWrappers(rhsPath), elementPathsLength)) {
      return rhsPath;
    }
  }
  return false;
}

const index = declare(api => {
  const {
    types: t,
    assertVersion
  } = api;
  assertVersion("^7.0.0-0 || ^8.0.0");
  return {
    name: "plugin-bugfix-safari-rest-destructuring-rhs-array",
    visitor: {
      ArrayPattern(path) {
        const rhs = shouldTransform(path);
        if (rhs) {
          path.replaceWith(t.arrayPattern([path.node]));
          rhs.replaceWith(t.arrayExpression([rhs.node]));
        }
      }
    }
  };
});

export { index as default };
