// @babel/preset-env/lib/index.js
import semver from 'semver';
import getTargets, { getInclusionReasons, filterItems, prettifyTargets } from '@babel/helper-compilation-targets';
import transformAsyncGeneratorFunctions from '@babel/plugin-transform-async-generator-functions';
import transformAsyncToGenerator from '@babel/plugin-transform-async-to-generator';
import transformArrowFunctions from '@babel/plugin-transform-arrow-functions';
import transformBlockScopedFunctions from '@babel/plugin-transform-block-scoped-functions';
import transformBlockScoping from '@babel/plugin-transform-block-scoping';
import transformClasses from '@babel/plugin-transform-classes';
import transformClassProperties from '@babel/plugin-transform-class-properties';
import transformClassStaticBlock from '@babel/plugin-transform-class-static-block';
import transformComputedProperties from '@babel/plugin-transform-computed-properties';
import transformDestructuring from '@babel/plugin-transform-destructuring';
import transformDotallRegex from '@babel/plugin-transform-dotall-regex';
import transformDuplicateKeys from '@babel/plugin-transform-duplicate-keys';
import transformDuplicateNamedCapturingGroupsRegex from '@babel/plugin-transform-duplicate-named-capturing-groups-regex';
import transformDynamicImport from '@babel/plugin-transform-dynamic-import';
import transformExplicitResourceManagement from '@babel/plugin-transform-explicit-resource-management';
import transformExponentialOperator from '@babel/plugin-transform-exponentiation-operator';
import transformExportNamespaceFrom from '@babel/plugin-transform-export-namespace-from';
import transformForOf from '@babel/plugin-transform-for-of';
import transformFunctionName from '@babel/plugin-transform-function-name';
import transformJsonStrings from '@babel/plugin-transform-json-strings';
import transformLiterals from '@babel/plugin-transform-literals';
import transformLogicalAssignmentOperators from '@babel/plugin-transform-logical-assignment-operators';
import transformMemberExpressionLiterals from '@babel/plugin-transform-member-expression-literals';
import transformModulesAmd from '@babel/plugin-transform-modules-amd';
import transformModulesCommonjs from '@babel/plugin-transform-modules-commonjs';
import transformModulesSystemjs from '@babel/plugin-transform-modules-systemjs';
import transformModulesUmd from '@babel/plugin-transform-modules-umd';
import transformNamedCapturingGroupsRegex from '@babel/plugin-transform-named-capturing-groups-regex';
import transformNewTarget from '@babel/plugin-transform-new-target';
import transformNullishCoalescingOperator from '@babel/plugin-transform-nullish-coalescing-operator';
import transformNumericSeparator from '@babel/plugin-transform-numeric-separator';
import transformObjectRestSpread from '@babel/plugin-transform-object-rest-spread';
import transformObjectSuper from '@babel/plugin-transform-object-super';
import transformOptionalCatchBinding from '@babel/plugin-transform-optional-catch-binding';
import transformOptionalChaining from '@babel/plugin-transform-optional-chaining';
import transformParameters from '@babel/plugin-transform-parameters';
import transformPrivateMethods from '@babel/plugin-transform-private-methods';
import transformPrivatePropertyInObject from '@babel/plugin-transform-private-property-in-object';
import transformPropertyLiterals from '@babel/plugin-transform-property-literals';
import transformRegenerator from '@babel/plugin-transform-regenerator';
import transformRegExpModifiers from '@babel/plugin-transform-regexp-modifiers';
import transformReservedWords from '@babel/plugin-transform-reserved-words';
import transformShorthandProperties from '@babel/plugin-transform-shorthand-properties';
import transformSpread from '@babel/plugin-transform-spread';
import transformStickyRegex from '@babel/plugin-transform-sticky-regex';
import transformTemplateLiterals from '@babel/plugin-transform-template-literals';
import transformTypeofSymbol from '@babel/plugin-transform-typeof-symbol';
import transformUnicodeEscapes from '@babel/plugin-transform-unicode-escapes';
import transformUnicodePropertyRegex from '@babel/plugin-transform-unicode-property-regex';
import transformUnicodeRegex from '@babel/plugin-transform-unicode-regex';
import transformUnicodeSetsRegex from '@babel/plugin-transform-unicode-sets-regex';
import bugfixAsyncArrowsInClass from '@babel/preset-modules/lib/plugins/transform-async-arrows-in-class/index.js';
import bugfixEdgeDefaultParameters from '@babel/preset-modules/lib/plugins/transform-edge-default-parameters/index.js';
import bugfixEdgeFunctionName from '@babel/preset-modules/lib/plugins/transform-edge-function-name/index.js';
import bugfixFirefoxClassInComputedKey from '@babel/plugin-bugfix-firefox-class-in-computed-class-key';
import bugfixTaggedTemplateCaching from '@babel/preset-modules/lib/plugins/transform-tagged-template-caching/index.js';
import bugfixSafariBlockShadowing from '@babel/preset-modules/lib/plugins/transform-safari-block-shadowing/index.js';
import bugfixSafariForShadowing from '@babel/preset-modules/lib/plugins/transform-safari-for-shadowing/index.js';
import bugfixSafariIdDestructuringCollisionInFunctionExpression from '@babel/plugin-bugfix-safari-id-destructuring-collision-in-function-expression';
import bugfixSafariRestDestructuringRhsArray from '@babel/plugin-bugfix-safari-rest-destructuring-rhs-array';
import bugfixSafariClassFieldInitializerScope from '@babel/plugin-bugfix-safari-class-field-initializer-scope';
import bugfixV8SpreadParametersInOptionalChaining from '@babel/plugin-bugfix-v8-spread-parameters-in-optional-chaining';
import bugfixV8StaticClassFieldsRedefineReadonly from '@babel/plugin-bugfix-v8-static-class-fields-redefine-readonly';
import originalPlugins from '@babel/compat-data/plugins' with { type: 'json' };
import originalPluginsBugfixes from '@babel/compat-data/plugin-bugfixes' with { type: 'json' };
import originalOverlappingPlugins from '@babel/compat-data/overlapping-plugins' with { type: 'json' };
import { OptionValidator } from '@babel/helper-validator-option';
import pluginCoreJS3 from 'babel-plugin-polyfill-corejs3';
import { declarePreset } from '@babel/helper-plugin-utils';

const logPlugin = (item, targetVersions, list) => {
  const filteredList = getInclusionReasons(item, targetVersions, list);
  const support = list[item];
  if (!support) {
    console.log(`  ${item}`);
    return;
  }
  let formattedTargets = `{`;
  let first = true;
  for (const target of Object.keys(filteredList)) {
    if (!first) formattedTargets += `,`;
    first = false;
    formattedTargets += ` ${target}`;
    if (support[target]) formattedTargets += ` < ${support[target]}`;
  }
  formattedTargets += ` }`;
  console.log(`  ${item} ${formattedTargets}`);
};

const availablePlugins = {
  "bugfix/transform-async-arrows-in-class": () => bugfixAsyncArrowsInClass,
  "bugfix/transform-edge-default-parameters": () => bugfixEdgeDefaultParameters,
  "bugfix/transform-edge-function-name": () => bugfixEdgeFunctionName,
  "bugfix/transform-firefox-class-in-computed-class-key": () => bugfixFirefoxClassInComputedKey,
  "bugfix/transform-safari-block-shadowing": () => bugfixSafariBlockShadowing,
  "bugfix/transform-safari-class-field-initializer-scope": () => bugfixSafariClassFieldInitializerScope,
  "bugfix/transform-safari-for-shadowing": () => bugfixSafariForShadowing,
  "bugfix/transform-safari-id-destructuring-collision-in-function-expression": () => bugfixSafariIdDestructuringCollisionInFunctionExpression,
  "bugfix/transform-safari-rest-destructuring-rhs-array": () => bugfixSafariRestDestructuringRhsArray,
  "bugfix/transform-tagged-template-caching": () => bugfixTaggedTemplateCaching,
  "bugfix/transform-v8-spread-parameters-in-optional-chaining": () => bugfixV8SpreadParametersInOptionalChaining,
  "bugfix/transform-v8-static-class-fields-redefine-readonly": () => bugfixV8StaticClassFieldsRedefineReadonly,
  "transform-arrow-functions": () => transformArrowFunctions,
  "transform-async-generator-functions": () => transformAsyncGeneratorFunctions,
  "transform-async-to-generator": () => transformAsyncToGenerator,
  "transform-block-scoped-functions": () => transformBlockScopedFunctions,
  "transform-block-scoping": () => transformBlockScoping,
  "transform-class-properties": () => transformClassProperties,
  "transform-class-static-block": () => transformClassStaticBlock,
  "transform-classes": () => transformClasses,
  "transform-computed-properties": () => transformComputedProperties,
  "transform-destructuring": () => transformDestructuring,
  "transform-dotall-regex": () => transformDotallRegex,
  "transform-duplicate-keys": () => transformDuplicateKeys,
  "transform-duplicate-named-capturing-groups-regex": () => transformDuplicateNamedCapturingGroupsRegex,
  "transform-dynamic-import": () => transformDynamicImport,
  "transform-explicit-resource-management": () => transformExplicitResourceManagement,
  "transform-exponentiation-operator": () => transformExponentialOperator,
  "transform-export-namespace-from": () => transformExportNamespaceFrom,
  "transform-for-of": () => transformForOf,
  "transform-function-name": () => transformFunctionName,
  "transform-json-strings": () => transformJsonStrings,
  "transform-literals": () => transformLiterals,
  "transform-logical-assignment-operators": () => transformLogicalAssignmentOperators,
  "transform-member-expression-literals": () => transformMemberExpressionLiterals,
  "transform-modules-amd": () => transformModulesAmd,
  "transform-modules-commonjs": () => transformModulesCommonjs,
  "transform-modules-systemjs": () => transformModulesSystemjs,
  "transform-modules-umd": () => transformModulesUmd,
  "transform-named-capturing-groups-regex": () => transformNamedCapturingGroupsRegex,
  "transform-new-target": () => transformNewTarget,
  "transform-nullish-coalescing-operator": () => transformNullishCoalescingOperator,
  "transform-numeric-separator": () => transformNumericSeparator,
  "transform-object-rest-spread": () => transformObjectRestSpread,
  "transform-object-super": () => transformObjectSuper,
  "transform-optional-catch-binding": () => transformOptionalCatchBinding,
  "transform-optional-chaining": () => transformOptionalChaining,
  "transform-parameters": () => transformParameters,
  "transform-private-methods": () => transformPrivateMethods,
  "transform-private-property-in-object": () => transformPrivatePropertyInObject,
  "transform-property-literals": () => transformPropertyLiterals,
  "transform-regenerator": () => transformRegenerator,
  "transform-regexp-modifiers": () => transformRegExpModifiers,
  "transform-reserved-words": () => transformReservedWords,
  "transform-shorthand-properties": () => transformShorthandProperties,
  "transform-spread": () => transformSpread,
  "transform-sticky-regex": () => transformStickyRegex,
  "transform-template-literals": () => transformTemplateLiterals,
  "transform-typeof-symbol": () => transformTypeofSymbol,
  "transform-unicode-escapes": () => transformUnicodeEscapes,
  "transform-unicode-property-regex": () => transformUnicodePropertyRegex,
  "transform-unicode-regex": () => transformUnicodeRegex,
  "transform-unicode-sets-regex": () => transformUnicodeSetsRegex
};
const minVersions = {};

function addProposalSyntaxPlugins(items, proposalSyntaxPlugins) {
  proposalSyntaxPlugins.forEach(plugin => {
    items.add(plugin);
  });
}
function removeUnnecessaryItems(items, overlapping) {
  items.forEach(item => {
    overlapping[item]?.forEach(name => items.delete(name));
  });
}
function removeUnsupportedItems(items, babelVersion) {
  items.forEach(item => {
    if (Object.hasOwn(minVersions, item) && semver.lt(babelVersion, minVersions[item])) {
      items.delete(item);
    }
  });
}

const moduleTransformations = {
  amd: "transform-modules-amd",
  commonjs: "transform-modules-commonjs",
  cjs: "transform-modules-commonjs",
  systemjs: "transform-modules-systemjs",
  umd: "transform-modules-umd"
};

const corejs3Polyfills = {
	"es.symbol": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "15",
	electron: "0.37",
	firefox: "51",
	"firefox-android": "51",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "10.0",
	samsung: "5.0"
},
	"es.symbol.description": {
	android: "70",
	bun: "0.1.1",
	chrome: "70",
	"chrome-android": "70",
	deno: "1.0",
	edge: "79",
	electron: "5.0",
	firefox: "63",
	"firefox-android": "63",
	ios: "12.2",
	node: "11.0",
	oculus: "6.0",
	opera: "57",
	"opera-android": "49",
	opera_mobile: "49",
	quest: "6.0",
	rhino: "1.8.0",
	safari: "12.1",
	samsung: "10.0"
},
	"es.symbol.async-dispose": {
	android: "127",
	bun: "1.0.23",
	chrome: "127",
	"chrome-android": "127",
	deno: "1.38",
	edge: "127",
	electron: "32.0",
	firefox: "135",
	"firefox-android": "135",
	node: "20.5.0",
	oculus: "35.0",
	opera: "113",
	"opera-android": "84",
	opera_mobile: "84",
	quest: "35.0",
	samsung: "28.0"
},
	"es.symbol.async-iterator": {
	android: "63",
	bun: "0.1.1",
	chrome: "63",
	"chrome-android": "63",
	deno: "1.0",
	edge: "79",
	electron: "3.0",
	firefox: "55",
	"firefox-android": "55",
	ios: "12.0",
	node: "10.0",
	oculus: "5.0",
	opera: "50",
	"opera-android": "46",
	opera_mobile: "46",
	quest: "5.0",
	safari: "12.0",
	samsung: "8.0"
},
	"es.symbol.dispose": {
	android: "125",
	bun: "1.0.23",
	chrome: "125",
	"chrome-android": "125",
	deno: "1.38",
	edge: "125",
	electron: "31.0",
	firefox: "135",
	"firefox-android": "135",
	node: "20.5.0",
	oculus: "34.0",
	opera: "111",
	"opera-android": "83",
	opera_mobile: "83",
	quest: "34.0",
	samsung: "27.0"
},
	"es.symbol.has-instance": {
	android: "50",
	bun: "0.1.1",
	chrome: "50",
	"chrome-android": "50",
	deno: "1.0",
	edge: "15",
	electron: "1.1",
	firefox: "49",
	"firefox-android": "49",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "37",
	"opera-android": "37",
	opera_mobile: "37",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.symbol.is-concat-spreadable": {
	android: "48",
	bun: "0.1.1",
	chrome: "48",
	"chrome-android": "48",
	deno: "1.0",
	edge: "15",
	electron: "0.37",
	firefox: "48",
	"firefox-android": "48",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "35",
	"opera-android": "35",
	opera_mobile: "35",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.symbol.iterator": {
	android: "41",
	bun: "0.1.1",
	chrome: "41",
	"chrome-android": "41",
	deno: "1.0",
	edge: "13",
	electron: "0.21",
	firefox: "36",
	"firefox-android": "36",
	hermes: "0.1",
	ios: "9.0",
	node: "1.0",
	oculus: "3.0",
	opera: "28",
	"opera-android": "28",
	opera_mobile: "28",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.4"
},
	"es.symbol.match": {
	android: "50",
	bun: "0.1.1",
	chrome: "50",
	"chrome-android": "50",
	deno: "1.0",
	edge: "79",
	electron: "1.1",
	firefox: "40",
	"firefox-android": "40",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "37",
	"opera-android": "37",
	opera_mobile: "37",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.symbol.match-all": {
	android: "73",
	bun: "0.1.1",
	chrome: "73",
	"chrome-android": "73",
	deno: "1.0",
	edge: "79",
	electron: "5.0",
	firefox: "67",
	"firefox-android": "67",
	hermes: "0.6",
	ios: "13.0",
	node: "12.0",
	oculus: "6.0",
	opera: "60",
	"opera-android": "52",
	opera_mobile: "52",
	quest: "6.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "13",
	samsung: "11.0"
},
	"es.symbol.replace": {
	android: "50",
	bun: "0.1.1",
	chrome: "50",
	"chrome-android": "50",
	deno: "1.0",
	edge: "79",
	electron: "1.1",
	firefox: "49",
	"firefox-android": "49",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "37",
	"opera-android": "37",
	opera_mobile: "37",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.symbol.search": {
	android: "50",
	bun: "0.1.1",
	chrome: "50",
	"chrome-android": "50",
	deno: "1.0",
	edge: "79",
	electron: "1.1",
	firefox: "49",
	"firefox-android": "49",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "37",
	"opera-android": "37",
	opera_mobile: "37",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.symbol.species": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "13",
	electron: "1.2",
	firefox: "41",
	"firefox-android": "41",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.symbol.split": {
	android: "50",
	bun: "0.1.1",
	chrome: "50",
	"chrome-android": "50",
	deno: "1.0",
	edge: "79",
	electron: "1.1",
	firefox: "49",
	"firefox-android": "49",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "37",
	"opera-android": "37",
	opera_mobile: "37",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.symbol.to-primitive": {
	android: "47",
	bun: "0.1.1",
	chrome: "47",
	"chrome-android": "47",
	deno: "1.0",
	edge: "15",
	electron: "0.36",
	firefox: "44",
	"firefox-android": "44",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "34",
	"opera-android": "34",
	opera_mobile: "34",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.symbol.to-string-tag": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "15",
	electron: "0.37",
	firefox: "51",
	"firefox-android": "51",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.symbol.unscopables": {
	android: "41",
	bun: "0.1.1",
	chrome: "41",
	"chrome-android": "41",
	deno: "1.0",
	edge: "13",
	electron: "0.21",
	firefox: "48",
	"firefox-android": "48",
	ios: "9.0",
	node: "1.0",
	oculus: "3.0",
	opera: "28",
	"opera-android": "28",
	opera_mobile: "28",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.4"
},
	"es.error.cause": {
	android: "94",
	bun: "0.1.1",
	chrome: "94",
	"chrome-android": "94",
	deno: "1.14",
	edge: "94",
	electron: "15.0",
	firefox: "91",
	"firefox-android": "91",
	hermes: "0.8",
	ios: "15.0",
	node: "16.11",
	oculus: "18.0",
	opera: "80",
	"opera-android": "66",
	opera_mobile: "66",
	quest: "18.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "15.0",
	samsung: "17.0"
},
	"es.error.is-error": {
	android: "134",
	chrome: "134",
	"chrome-android": "134",
	deno: "2.2",
	edge: "134",
	electron: "35.0",
	firefox: "138",
	"firefox-android": "138",
	node: "24.3",
	oculus: "38.0",
	opera: "119",
	"opera-android": "88",
	opera_mobile: "88",
	quest: "38.0",
	rhino: "1.9.0",
	samsung: "29.0"
},
	"es.error.to-string": {
	android: "4.4.3",
	bun: "0.1.1",
	chrome: "33",
	"chrome-android": "33",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "11",
	"firefox-android": "11",
	hermes: "0.1",
	ie: "9",
	ios: "9.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "20",
	"opera-android": "20",
	opera_mobile: "20",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "8.0",
	samsung: "2.0"
},
	"es.aggregate-error": {
	android: "85",
	bun: "0.1.1",
	chrome: "85",
	"chrome-android": "85",
	deno: "1.2",
	edge: "85",
	electron: "10.0",
	firefox: "79",
	"firefox-android": "79",
	hermes: "0.13",
	ios: "14.0",
	node: "15.0",
	oculus: "12.0",
	opera: "71",
	"opera-android": "60",
	opera_mobile: "60",
	quest: "12.0",
	"react-native": "0.72",
	rhino: "1.8.0",
	safari: "14.0",
	samsung: "14.0"
},
	"es.aggregate-error.cause": {
	android: "94",
	bun: "0.1.1",
	chrome: "94",
	"chrome-android": "94",
	deno: "1.14",
	edge: "94",
	electron: "15.0",
	firefox: "91",
	"firefox-android": "91",
	hermes: "0.13",
	ios: "15.0",
	node: "16.11",
	oculus: "18.0",
	opera: "80",
	"opera-android": "66",
	opera_mobile: "66",
	quest: "18.0",
	"react-native": "0.72",
	rhino: "1.8.0",
	safari: "15.0",
	samsung: "17.0"
},
	"es.suppressed-error.constructor": {
	android: "136",
	bun: "1.2.15",
	chrome: "136",
	"chrome-android": "136",
	deno: "2.2.10",
	edge: "136",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "90",
	opera_mobile: "90",
	quest: "39.0",
	samsung: "29.0"
},
	"es.array.at": {
	android: "92",
	bun: "0.1.1",
	chrome: "92",
	"chrome-android": "92",
	deno: "1.12",
	edge: "92",
	electron: "14.0",
	firefox: "90",
	"firefox-android": "90",
	hermes: "0.13",
	ios: "15.4",
	node: "16.6",
	oculus: "17.0",
	opera: "78",
	"opera-android": "65",
	opera_mobile: "65",
	quest: "17.0",
	"react-native": "0.71",
	rhino: "1.7.15",
	safari: "15.4",
	samsung: "16.0"
},
	"es.array.concat": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "15",
	electron: "1.2",
	firefox: "48",
	"firefox-android": "48",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.array.copy-within": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "12",
	electron: "0.31",
	firefox: "48",
	"firefox-android": "48",
	ios: "9.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	rhino: "1.8.0",
	safari: "9.0",
	samsung: "5.0"
},
	"es.array.every": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "26",
	"chrome-android": "26",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "8.0",
	node: "0.11.0",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "1.5"
},
	"es.array.fill": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "12",
	electron: "0.31",
	firefox: "48",
	"firefox-android": "48",
	ios: "9.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	rhino: "1.8.0",
	safari: "9.0",
	samsung: "5.0"
},
	"es.array.filter": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "15",
	electron: "1.2",
	firefox: "48",
	"firefox-android": "48",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.array.find": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "48",
	"firefox-android": "48",
	ios: "9.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	rhino: "1.8.0",
	safari: "9.0",
	samsung: "5.0"
},
	"es.array.find-index": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "48",
	"firefox-android": "48",
	ios: "9.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	rhino: "1.8.0",
	safari: "9.0",
	samsung: "5.0"
},
	"es.array.find-last": {
	android: "97",
	bun: "0.1.1",
	chrome: "97",
	"chrome-android": "97",
	deno: "1.16",
	edge: "97",
	electron: "17.0",
	firefox: "104",
	"firefox-android": "104",
	hermes: "0.11",
	ios: "15.4",
	node: "18.0",
	oculus: "20.0",
	opera: "83",
	"opera-android": "68",
	opera_mobile: "68",
	quest: "20.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "15.4",
	samsung: "18.0"
},
	"es.array.find-last-index": {
	android: "97",
	bun: "0.1.1",
	chrome: "97",
	"chrome-android": "97",
	deno: "1.16",
	edge: "97",
	electron: "17.0",
	firefox: "104",
	"firefox-android": "104",
	hermes: "0.11",
	ios: "15.4",
	node: "18.0",
	oculus: "20.0",
	opera: "83",
	"opera-android": "68",
	opera_mobile: "68",
	quest: "20.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "15.4",
	samsung: "18.0"
},
	"es.array.flat": {
	android: "69",
	bun: "0.1.1",
	chrome: "69",
	"chrome-android": "69",
	deno: "1.0",
	edge: "79",
	electron: "4.0",
	firefox: "62",
	"firefox-android": "62",
	hermes: "0.4",
	ios: "12.0",
	node: "11.0",
	oculus: "6.0",
	opera: "56",
	"opera-android": "48",
	opera_mobile: "48",
	quest: "6.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "12.0",
	samsung: "10.0"
},
	"es.array.flat-map": {
	android: "69",
	bun: "0.1.1",
	chrome: "69",
	"chrome-android": "69",
	deno: "1.0",
	edge: "79",
	electron: "4.0",
	firefox: "62",
	"firefox-android": "62",
	hermes: "0.4",
	ios: "12.0",
	node: "11.0",
	oculus: "6.0",
	opera: "56",
	"opera-android": "48",
	opera_mobile: "48",
	quest: "6.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "12.0",
	samsung: "10.0"
},
	"es.array.for-each": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "26",
	"chrome-android": "26",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "8.0",
	node: "0.11.0",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "1.5"
},
	"es.array.from": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "15",
	electron: "1.2",
	firefox: "53",
	"firefox-android": "53",
	hermes: "0.13",
	ios: "9.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.73",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "5.0"
},
	"es.array.includes": {
	android: "53",
	bun: "0.1.1",
	chrome: "53",
	"chrome-android": "53",
	deno: "1.0",
	edge: "14",
	electron: "1.4",
	firefox: "102",
	"firefox-android": "102",
	ios: "10.0",
	node: "7.0",
	oculus: "3.0",
	opera: "40",
	"opera-android": "40",
	opera_mobile: "40",
	quest: "3.0",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "6.0"
},
	"es.array.index-of": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "12",
	electron: "1.2",
	firefox: "47",
	"firefox-android": "47",
	hermes: "0.1",
	ie: "9",
	ios: "8.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "5.0"
},
	"es.array.is-array": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "3.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "4.0",
	samsung: "1.0"
},
	"es.array.iterator": {
	android: "66",
	bun: "0.1.1",
	chrome: "66",
	"chrome-android": "66",
	deno: "1.0",
	edge: "15",
	electron: "3.0",
	firefox: "60",
	"firefox-android": "60",
	ios: "10.0",
	node: "10.0",
	oculus: "5.0",
	opera: "53",
	"opera-android": "47",
	opera_mobile: "47",
	quest: "5.0",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "9.0"
},
	"es.array.join": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "26",
	"chrome-android": "26",
	deno: "1.0",
	edge: "13",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.0",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "1.5"
},
	"es.array.last-index-of": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "12",
	electron: "1.2",
	firefox: "47",
	"firefox-android": "47",
	hermes: "0.1",
	ie: "9",
	ios: "8.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "5.0"
},
	"es.array.map": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "13",
	electron: "1.2",
	firefox: "50",
	"firefox-android": "50",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.array.of": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "9.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "5.0"
},
	"es.array.push": {
	android: "122",
	bun: "0.1.1",
	chrome: "122",
	"chrome-android": "122",
	deno: "1.41.3",
	edge: "122",
	electron: "29.0",
	firefox: "55",
	"firefox-android": "55",
	hermes: "0.2",
	ios: "16.0",
	node: "22.0",
	oculus: "32.0",
	opera: "108",
	"opera-android": "81",
	opera_mobile: "81",
	quest: "32.0",
	"react-native": "0.69",
	safari: "16.0",
	samsung: "26.0"
},
	"es.array.reduce": {
	android: "83",
	bun: "0.1.1",
	chrome: "83",
	"chrome-android": "83",
	deno: "1.0",
	edge: "12",
	electron: "9.0",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "8.0",
	node: "6.0",
	oculus: "10.0",
	opera: "69",
	"opera-android": "59",
	opera_mobile: "59",
	quest: "10.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "13.0"
},
	"es.array.reduce-right": {
	android: "83",
	bun: "0.1.1",
	chrome: "83",
	"chrome-android": "83",
	deno: "1.0",
	edge: "12",
	electron: "9.0",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "8.0",
	node: "6.0",
	oculus: "10.0",
	opera: "69",
	"opera-android": "59",
	opera_mobile: "59",
	quest: "10.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "13.0"
},
	"es.array.reverse": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "1",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "1",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "5.5",
	ios: "12.2",
	node: "0.0.3",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "12.0.2",
	samsung: "1.0"
},
	"es.array.slice": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "13",
	electron: "1.2",
	firefox: "48",
	"firefox-android": "48",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.array.some": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "26",
	"chrome-android": "26",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "8.0",
	node: "0.11.0",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "1.5"
},
	"es.array.sort": {
	android: "70",
	bun: "0.1.1",
	chrome: "70",
	"chrome-android": "70",
	deno: "1.0",
	edge: "79",
	electron: "5.0",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.10",
	ios: "12.0",
	node: "11.0",
	oculus: "6.0",
	opera: "57",
	"opera-android": "49",
	opera_mobile: "49",
	quest: "6.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "12.0",
	samsung: "10.0"
},
	"es.array.species": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "13",
	electron: "1.2",
	firefox: "48",
	"firefox-android": "48",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	rhino: "1.7.15",
	safari: "10.0",
	samsung: "5.0"
},
	"es.array.splice": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "13",
	electron: "1.2",
	firefox: "49",
	"firefox-android": "49",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.array.to-reversed": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	hermes: "0.13",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	"react-native": "0.74",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"es.array.to-sorted": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"es.array.to-spliced": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	hermes: "0.13",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	"react-native": "0.74",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"es.array.unscopables.flat": {
	android: "73",
	bun: "0.1.1",
	chrome: "73",
	"chrome-android": "73",
	deno: "1.0",
	edge: "79",
	electron: "5.0",
	firefox: "67",
	"firefox-android": "67",
	ios: "13.0",
	node: "12.0",
	oculus: "6.0",
	opera: "60",
	"opera-android": "52",
	opera_mobile: "52",
	quest: "6.0",
	rhino: "1.8.0",
	safari: "13",
	samsung: "11.0"
},
	"es.array.unscopables.flat-map": {
	android: "73",
	bun: "0.1.1",
	chrome: "73",
	"chrome-android": "73",
	deno: "1.0",
	edge: "79",
	electron: "5.0",
	firefox: "67",
	"firefox-android": "67",
	ios: "13.0",
	node: "12.0",
	oculus: "6.0",
	opera: "60",
	"opera-android": "52",
	opera_mobile: "52",
	quest: "6.0",
	rhino: "1.8.0",
	safari: "13",
	samsung: "11.0"
},
	"es.array.unshift": {
	android: "71",
	bun: "0.1.1",
	chrome: "71",
	"chrome-android": "71",
	deno: "1.0",
	edge: "12",
	electron: "5.0",
	firefox: "23",
	"firefox-android": "23",
	hermes: "0.1",
	ie: "9",
	ios: "16.0",
	node: "12.0",
	oculus: "6.0",
	opera: "58",
	"opera-android": "50",
	opera_mobile: "50",
	quest: "6.0",
	"react-native": "0.69",
	safari: "16.0",
	samsung: "10.0"
},
	"es.array.with": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "140",
	"firefox-android": "140",
	hermes: "0.13",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	"react-native": "0.74",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"es.array-buffer.constructor": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "28",
	"chrome-android": "28",
	deno: "1.0",
	edge: "14",
	electron: "0.20",
	firefox: "44",
	"firefox-android": "44",
	hermes: "0.1",
	ios: "12.0",
	node: "0.11.1",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "12.0",
	samsung: "1.5"
},
	"es.array-buffer.is-view": {
	android: "4.4.3",
	bun: "0.1.1",
	chrome: "32",
	"chrome-android": "32",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "29",
	"firefox-android": "29",
	hermes: "0.1",
	ie: "11",
	ios: "8.0",
	node: "0.11.9",
	oculus: "3.0",
	opera: "19",
	"opera-android": "19",
	opera_mobile: "19",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "7.1",
	samsung: "2.0"
},
	"es.array-buffer.slice": {
	android: "4.4.3",
	bun: "0.1.1",
	chrome: "31",
	"chrome-android": "31",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "46",
	"firefox-android": "46",
	hermes: "0.1",
	ie: "11",
	ios: "12.2",
	node: "0.11.8",
	oculus: "3.0",
	opera: "18",
	"opera-android": "18",
	opera_mobile: "18",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "12.1",
	samsung: "2.0"
},
	"es.data-view": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "26",
	"chrome-android": "26",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "15",
	"firefox-android": "15",
	hermes: "0.1",
	ie: "10",
	ios: "8.0",
	node: "0.11.0",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "1.5"
},
	"es.data-view.get-float16": {
	android: "135",
	bun: "1.1.23",
	chrome: "135",
	"chrome-android": "135",
	deno: "1.43",
	edge: "135",
	electron: "36.0",
	firefox: "129",
	"firefox-android": "129",
	ios: "18.2",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "18.2",
	samsung: "29.0"
},
	"es.data-view.set-float16": {
	android: "135",
	bun: "1.1.23",
	chrome: "135",
	"chrome-android": "135",
	deno: "1.43",
	edge: "135",
	electron: "36.0",
	firefox: "129",
	"firefox-android": "129",
	ios: "18.2",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "18.2",
	samsung: "29.0"
},
	"es.array-buffer.detached": {
	android: "114",
	bun: "1.0.19",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "122",
	"firefox-android": "122",
	ios: "17.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	rhino: "1.9.0",
	safari: "17.4",
	samsung: "23.0"
},
	"es.array-buffer.transfer": {
	android: "114",
	bun: "1.0.19",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "122",
	"firefox-android": "122",
	ios: "17.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	rhino: "1.9.0",
	safari: "17.4",
	samsung: "23.0"
},
	"es.array-buffer.transfer-to-fixed-length": {
	android: "114",
	bun: "1.0.19",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "122",
	"firefox-android": "122",
	ios: "17.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	rhino: "1.9.0",
	safari: "17.4",
	samsung: "23.0"
},
	"es.date.get-year": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "1",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "1",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "1.0",
	node: "0.0.3",
	oculus: "3.0",
	opera: "3",
	"opera-android": "3",
	opera_mobile: "3",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "1",
	samsung: "1.0"
},
	"es.date.now": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "3.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "4.0",
	samsung: "1.0"
},
	"es.date.set-year": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "1",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "1",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "3",
	ios: "1.0",
	node: "0.0.3",
	oculus: "3.0",
	opera: "3",
	"opera-android": "3",
	opera_mobile: "3",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "1",
	samsung: "1.0"
},
	"es.date.to-gmt-string": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "1",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "1",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "3",
	ios: "1.0",
	node: "0.0.3",
	oculus: "3.0",
	opera: "3",
	"opera-android": "3",
	opera_mobile: "3",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "1",
	samsung: "1.0"
},
	"es.date.to-iso-string": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "26",
	"chrome-android": "26",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "7",
	"firefox-android": "7",
	hermes: "0.1",
	ie: "9",
	ios: "8.0",
	node: "0.11.0",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "1.5"
},
	"es.date.to-json": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "26",
	"chrome-android": "26",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "10.0",
	node: "0.11.0",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "1.5"
},
	"es.date.to-primitive": {
	android: "47",
	bun: "0.1.1",
	chrome: "47",
	"chrome-android": "47",
	deno: "1.0",
	edge: "15",
	electron: "0.36",
	firefox: "44",
	"firefox-android": "44",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "34",
	"opera-android": "34",
	opera_mobile: "34",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.date.to-string": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "3.1",
	samsung: "1.0"
},
	"es.disposable-stack.constructor": {
	android: "136",
	bun: "1.3.0",
	chrome: "136",
	"chrome-android": "136",
	deno: "2.2.10",
	edge: "136",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "90",
	opera_mobile: "90",
	quest: "39.0",
	samsung: "29.0"
},
	"es.escape": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "1",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "1",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "3",
	ios: "1.0",
	node: "0.0.3",
	oculus: "3.0",
	opera: "3",
	"opera-android": "3",
	opera_mobile: "3",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "1",
	samsung: "1.0"
},
	"es.function.bind": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "7",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "5.0",
	node: "0.1.101",
	oculus: "3.0",
	opera: "12",
	"opera-android": "12",
	opera_mobile: "12",
	phantom: "2.0",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "5.1",
	samsung: "1.0"
},
	"es.function.has-instance": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "15",
	electron: "1.2",
	firefox: "50",
	"firefox-android": "50",
	hermes: "0.1",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.function.name": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	hermes: "0.1",
	ios: "3.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "4.0",
	samsung: "1.0"
},
	"es.global-this": {
	android: "71",
	bun: "0.1.1",
	chrome: "71",
	"chrome-android": "71",
	deno: "1.0",
	edge: "79",
	electron: "5.0",
	firefox: "65",
	"firefox-android": "65",
	hermes: "0.2",
	ios: "12.2",
	node: "12.0",
	oculus: "6.0",
	opera: "58",
	"opera-android": "50",
	opera_mobile: "50",
	quest: "6.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "12.1",
	samsung: "10.0"
},
	"es.iterator.constructor": {
	android: "122",
	bun: "1.1.31",
	chrome: "122",
	"chrome-android": "122",
	deno: "1.38.1",
	edge: "122",
	electron: "29.0",
	firefox: "131",
	"firefox-android": "131",
	ios: "18.4",
	node: "22.0",
	oculus: "32.0",
	opera: "108",
	"opera-android": "81",
	opera_mobile: "81",
	quest: "32.0",
	safari: "18.4",
	samsung: "26.0"
},
	"es.iterator.concat": {
	bun: "1.3.7",
	firefox: "147",
	"firefox-android": "147"
},
	"es.iterator.dispose": {
	android: "136",
	bun: "1.3.0",
	chrome: "136",
	"chrome-android": "136",
	deno: "2.2.10",
	edge: "136",
	electron: "36.0",
	firefox: "135",
	"firefox-android": "135",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "90",
	opera_mobile: "90",
	quest: "39.0",
	samsung: "29.0"
},
	"es.iterator.drop": {
	android: "135",
	bun: "1.2.11",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"es.iterator.every": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"es.iterator.filter": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"es.iterator.find": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"es.iterator.flat-map": {
	android: "135",
	bun: "1.2.21",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.2",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.2",
	samsung: "29.0"
},
	"es.iterator.for-each": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"es.iterator.from": {
	android: "122",
	bun: "1.2.5",
	chrome: "122",
	"chrome-android": "122",
	deno: "1.38.1",
	edge: "122",
	electron: "29.0",
	firefox: "131",
	"firefox-android": "131",
	ios: "26.0",
	node: "22.0",
	oculus: "32.0",
	opera: "108",
	"opera-android": "81",
	opera_mobile: "81",
	quest: "32.0",
	safari: "26.0",
	samsung: "26.0"
},
	"es.iterator.map": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"es.iterator.reduce": {
	android: "135",
	bun: "1.2.11",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"es.iterator.some": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"es.iterator.take": {
	android: "135",
	bun: "1.2.11",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"es.iterator.to-array": {
	android: "122",
	bun: "1.1.31",
	chrome: "122",
	"chrome-android": "122",
	deno: "1.38.1",
	edge: "122",
	electron: "29.0",
	firefox: "131",
	"firefox-android": "131",
	ios: "18.4",
	node: "22.0",
	oculus: "32.0",
	opera: "108",
	"opera-android": "81",
	opera_mobile: "81",
	quest: "32.0",
	safari: "18.4",
	samsung: "26.0"
},
	"es.json.is-raw-json": {
	android: "114",
	bun: "1.1.43",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "135",
	"firefox-android": "135",
	ios: "18.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	safari: "18.4",
	samsung: "23.0"
},
	"es.json.parse": {
	android: "114",
	bun: "1.1.43",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "135",
	"firefox-android": "135",
	ios: "18.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	safari: "18.4",
	samsung: "23.0"
},
	"es.json.raw-json": {
	android: "114",
	bun: "1.1.43",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "135",
	"firefox-android": "135",
	ios: "18.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	safari: "18.4",
	samsung: "23.0"
},
	"es.json.stringify": {
	android: "114",
	bun: "1.1.43",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "135",
	"firefox-android": "135",
	ios: "18.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	safari: "18.4",
	samsung: "23.0"
},
	"es.json.to-string-tag": {
	android: "50",
	bun: "0.1.1",
	chrome: "50",
	"chrome-android": "50",
	deno: "1.0",
	edge: "15",
	electron: "1.1",
	firefox: "51",
	"firefox-android": "51",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "37",
	"opera-android": "37",
	opera_mobile: "37",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "10.0",
	samsung: "5.0"
},
	"es.map": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "15",
	electron: "1.2",
	firefox: "53",
	"firefox-android": "53",
	hermes: "0.13",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.73",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.map.group-by": {
	android: "117",
	bun: "1.1.2",
	chrome: "117",
	"chrome-android": "117",
	deno: "1.37",
	edge: "117",
	electron: "27.0",
	firefox: "119",
	"firefox-android": "119",
	ios: "18.0",
	node: "21.0",
	oculus: "30.0",
	opera: "103",
	"opera-android": "78",
	opera_mobile: "78",
	quest: "30.0",
	rhino: "1.8.0",
	safari: "18.0",
	samsung: "24.0"
},
	"es.map.get-or-insert": {
	android: "145",
	bun: "1.2.20",
	chrome: "145",
	"chrome-android": "145",
	edge: "145",
	electron: "41.0",
	firefox: "144",
	"firefox-android": "144",
	ios: "26.2",
	opera: "129",
	safari: "26.2"
},
	"es.map.get-or-insert-computed": {
	android: "145",
	bun: "1.2.20",
	chrome: "145",
	"chrome-android": "145",
	edge: "145",
	electron: "41.0",
	firefox: "144",
	"firefox-android": "144",
	ios: "26.2",
	opera: "129",
	safari: "26.2"
},
	"es.math.acosh": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "13",
	electron: "1.4",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	"react-native": "0.69",
	safari: "7.1",
	samsung: "6.0"
},
	"es.math.asinh": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "13",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.math.atanh": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "13",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.math.cbrt": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.math.clz32": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "31",
	"firefox-android": "31",
	hermes: "0.1",
	ios: "9.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.0"
},
	"es.math.cosh": {
	android: "39",
	bun: "0.1.1",
	chrome: "39",
	"chrome-android": "39",
	deno: "1.0",
	edge: "13",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "1.0",
	oculus: "3.0",
	opera: "26",
	"opera-android": "26",
	opera_mobile: "26",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.4"
},
	"es.math.expm1": {
	android: "39",
	bun: "0.1.1",
	chrome: "39",
	"chrome-android": "39",
	deno: "1.0",
	edge: "13",
	electron: "0.20",
	firefox: "46",
	"firefox-android": "46",
	hermes: "0.1",
	ios: "8.0",
	node: "1.0",
	oculus: "3.0",
	opera: "26",
	"opera-android": "26",
	opera_mobile: "26",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.4"
},
	"es.math.fround": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "26",
	"firefox-android": "26",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.math.f16round": {
	android: "135",
	bun: "1.1.23",
	chrome: "135",
	"chrome-android": "135",
	deno: "1.43",
	edge: "135",
	electron: "36.0",
	firefox: "129",
	"firefox-android": "129",
	ios: "18.2",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	rhino: "1.9.0",
	safari: "18.2",
	samsung: "29.0"
},
	"es.math.hypot": {
	android: "78",
	bun: "0.1.1",
	chrome: "78",
	"chrome-android": "78",
	deno: "1.0",
	edge: "12",
	electron: "7.0",
	firefox: "27",
	"firefox-android": "27",
	hermes: "0.1",
	ios: "8.0",
	node: "12.16",
	oculus: "8.0",
	opera: "65",
	"opera-android": "56",
	opera_mobile: "56",
	quest: "8.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "12.0"
},
	"es.math.imul": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "28",
	"chrome-android": "28",
	deno: "1.0",
	edge: "13",
	electron: "0.20",
	firefox: "20",
	"firefox-android": "20",
	hermes: "0.1",
	ios: "9.0",
	node: "0.11.1",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "1.5"
},
	"es.math.log10": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.math.log1p": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.math.log2": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.math.sign": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "9.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.0"
},
	"es.math.sinh": {
	android: "39",
	bun: "0.1.1",
	chrome: "39",
	"chrome-android": "39",
	deno: "1.0",
	edge: "13",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "1.0",
	oculus: "3.0",
	opera: "26",
	"opera-android": "26",
	opera_mobile: "26",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.4"
},
	"es.math.sum-precise": {
	bun: "1.2.18",
	firefox: "137",
	"firefox-android": "137",
	ios: "26.2",
	safari: "26.2"
},
	"es.math.tanh": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.math.to-string-tag": {
	android: "50",
	bun: "0.1.1",
	chrome: "50",
	"chrome-android": "50",
	deno: "1.0",
	edge: "15",
	electron: "1.1",
	firefox: "51",
	"firefox-android": "51",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "37",
	"opera-android": "37",
	opera_mobile: "37",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "10.0",
	samsung: "5.0"
},
	"es.math.trunc": {
	android: "38",
	bun: "0.1.1",
	chrome: "38",
	"chrome-android": "38",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "25",
	"opera-android": "25",
	opera_mobile: "25",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.number.constructor": {
	android: "41",
	bun: "0.1.1",
	chrome: "41",
	"chrome-android": "41",
	deno: "1.0",
	edge: "13",
	electron: "0.21",
	firefox: "46",
	"firefox-android": "46",
	hermes: "0.5",
	ios: "9.0",
	node: "1.0",
	oculus: "3.0",
	opera: "28",
	"opera-android": "28",
	opera_mobile: "28",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.4"
},
	"es.number.epsilon": {
	android: "37",
	bun: "0.1.1",
	chrome: "34",
	"chrome-android": "34",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "25",
	"firefox-android": "25",
	hermes: "0.1",
	ios: "9.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "21",
	"opera-android": "21",
	opera_mobile: "21",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "9.0",
	samsung: "2.0"
},
	"es.number.is-finite": {
	android: "4.1",
	bun: "0.1.1",
	chrome: "19",
	"chrome-android": "25",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "16",
	"firefox-android": "16",
	hermes: "0.1",
	ios: "9.0",
	node: "0.7.3",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "1.5"
},
	"es.number.is-integer": {
	android: "37",
	bun: "0.1.1",
	chrome: "34",
	"chrome-android": "34",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "16",
	"firefox-android": "16",
	hermes: "0.1",
	ios: "9.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "21",
	"opera-android": "21",
	opera_mobile: "21",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "2.0"
},
	"es.number.is-nan": {
	android: "4.1",
	bun: "0.1.1",
	chrome: "19",
	"chrome-android": "25",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "15",
	"firefox-android": "15",
	hermes: "0.1",
	ios: "9.0",
	node: "0.7.3",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "1.5"
},
	"es.number.is-safe-integer": {
	android: "37",
	bun: "0.1.1",
	chrome: "34",
	"chrome-android": "34",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "32",
	"firefox-android": "32",
	hermes: "0.1",
	ios: "9.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "21",
	"opera-android": "21",
	opera_mobile: "21",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "2.0"
},
	"es.number.max-safe-integer": {
	android: "37",
	bun: "0.1.1",
	chrome: "34",
	"chrome-android": "34",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "31",
	"firefox-android": "31",
	hermes: "0.1",
	ios: "9.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "21",
	"opera-android": "21",
	opera_mobile: "21",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "2.0"
},
	"es.number.min-safe-integer": {
	android: "37",
	bun: "0.1.1",
	chrome: "34",
	"chrome-android": "34",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "31",
	"firefox-android": "31",
	hermes: "0.1",
	ios: "9.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "21",
	"opera-android": "21",
	opera_mobile: "21",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "2.0"
},
	"es.number.parse-float": {
	android: "37",
	bun: "0.1.1",
	chrome: "35",
	"chrome-android": "35",
	deno: "1.0",
	edge: "79",
	electron: "0.20",
	firefox: "39",
	"firefox-android": "39",
	hermes: "0.1",
	ios: "11.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "22",
	"opera-android": "22",
	opera_mobile: "22",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "11.0",
	samsung: "3.0"
},
	"es.number.parse-int": {
	android: "37",
	bun: "0.1.1",
	chrome: "35",
	"chrome-android": "35",
	deno: "1.0",
	edge: "79",
	electron: "0.20",
	firefox: "39",
	"firefox-android": "39",
	hermes: "0.1",
	ios: "9.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "22",
	"opera-android": "22",
	opera_mobile: "22",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "9.0",
	samsung: "3.0"
},
	"es.number.to-exponential": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "18",
	electron: "1.2",
	firefox: "87",
	"firefox-android": "87",
	hermes: "0.1",
	ios: "11.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "11",
	samsung: "5.0"
},
	"es.number.to-fixed": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "26",
	"chrome-android": "26",
	deno: "1.0",
	edge: "79",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.0",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "1.5"
},
	"es.number.to-precision": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "26",
	"chrome-android": "26",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "8",
	ios: "8.0",
	node: "0.11.0",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "1.5"
},
	"es.object.assign": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "79",
	electron: "0.37",
	firefox: "36",
	"firefox-android": "36",
	hermes: "0.4",
	ios: "9.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "9.0",
	samsung: "5.0"
},
	"es.object.create": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "3.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "12",
	"opera-android": "12",
	opera_mobile: "12",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "4.0",
	samsung: "1.0"
},
	"es.object.define-getter": {
	android: "62",
	bun: "0.1.1",
	chrome: "62",
	"chrome-android": "62",
	deno: "1.0",
	edge: "16",
	electron: "3.0",
	firefox: "48",
	"firefox-android": "48",
	hermes: "0.1",
	ios: "8.0",
	node: "8.10",
	oculus: "5.0",
	opera: "49",
	"opera-android": "46",
	opera_mobile: "46",
	quest: "5.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "8.0"
},
	"es.object.define-properties": {
	android: "37",
	bun: "0.1.1",
	chrome: "37",
	"chrome-android": "37",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "5.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "12",
	"opera-android": "12",
	opera_mobile: "12",
	phantom: "2.0",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "5.1",
	samsung: "3.0"
},
	"es.object.define-property": {
	android: "37",
	bun: "0.1.1",
	chrome: "37",
	"chrome-android": "37",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "4",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "5.0",
	node: "0.11.15",
	oculus: "3.0",
	opera: "12",
	"opera-android": "12",
	opera_mobile: "12",
	phantom: "2.0",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "5.1",
	samsung: "3.0"
},
	"es.object.define-setter": {
	android: "62",
	bun: "0.1.1",
	chrome: "62",
	"chrome-android": "62",
	deno: "1.0",
	edge: "16",
	electron: "3.0",
	firefox: "48",
	"firefox-android": "48",
	hermes: "0.1",
	ios: "8.0",
	node: "8.10",
	oculus: "5.0",
	opera: "49",
	"opera-android": "46",
	opera_mobile: "46",
	quest: "5.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "8.0"
},
	"es.object.entries": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "14",
	electron: "1.4",
	firefox: "47",
	"firefox-android": "47",
	hermes: "0.1",
	ios: "10.3",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "10.1",
	samsung: "6.0"
},
	"es.object.freeze": {
	android: "44",
	bun: "0.1.1",
	chrome: "44",
	"chrome-android": "44",
	deno: "1.0",
	edge: "13",
	electron: "0.30",
	firefox: "35",
	"firefox-android": "35",
	hermes: "0.1",
	ios: "9.0",
	node: "3.0",
	oculus: "3.0",
	opera: "31",
	"opera-android": "31",
	opera_mobile: "31",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "4.0"
},
	"es.object.from-entries": {
	android: "73",
	bun: "0.1.1",
	chrome: "73",
	"chrome-android": "73",
	deno: "1.0",
	edge: "79",
	electron: "5.0",
	firefox: "63",
	"firefox-android": "63",
	hermes: "0.4",
	ios: "12.2",
	node: "12.0",
	oculus: "6.0",
	opera: "60",
	"opera-android": "52",
	opera_mobile: "52",
	quest: "6.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "12.1",
	samsung: "11.0"
},
	"es.object.get-own-property-descriptor": {
	android: "44",
	bun: "0.1.1",
	chrome: "44",
	"chrome-android": "44",
	deno: "1.0",
	edge: "13",
	electron: "0.30",
	firefox: "35",
	"firefox-android": "35",
	hermes: "0.1",
	ios: "9.0",
	node: "3.0",
	oculus: "3.0",
	opera: "31",
	"opera-android": "31",
	opera_mobile: "31",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "4.0"
},
	"es.object.get-own-property-descriptors": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "50",
	"firefox-android": "50",
	hermes: "0.6",
	ios: "10.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "10.0",
	samsung: "6.0"
},
	"es.object.get-own-property-names": {
	android: "40",
	bun: "0.1.1",
	chrome: "40",
	"chrome-android": "40",
	deno: "1.0",
	edge: "13",
	electron: "0.21",
	firefox: "34",
	"firefox-android": "34",
	hermes: "0.1",
	ios: "9.0",
	node: "1.0",
	oculus: "3.0",
	opera: "27",
	"opera-android": "27",
	opera_mobile: "27",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.4"
},
	"es.object.get-prototype-of": {
	android: "44",
	bun: "0.1.1",
	chrome: "44",
	"chrome-android": "44",
	deno: "1.0",
	edge: "13",
	electron: "0.30",
	firefox: "35",
	"firefox-android": "35",
	hermes: "0.1",
	ios: "9.0",
	node: "3.0",
	oculus: "3.0",
	opera: "31",
	"opera-android": "31",
	opera_mobile: "31",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "4.0"
},
	"es.object.group-by": {
	android: "117",
	bun: "1.1.2",
	chrome: "117",
	"chrome-android": "117",
	deno: "1.37",
	edge: "117",
	electron: "27.0",
	firefox: "119",
	"firefox-android": "119",
	ios: "18.0",
	node: "21.0",
	oculus: "30.0",
	opera: "103",
	"opera-android": "78",
	opera_mobile: "78",
	quest: "30.0",
	rhino: "1.8.0",
	safari: "18.0",
	samsung: "24.0"
},
	"es.object.has-own": {
	android: "93",
	bun: "0.1.1",
	chrome: "93",
	"chrome-android": "93",
	deno: "1.13",
	edge: "93",
	electron: "14.0",
	firefox: "92",
	"firefox-android": "92",
	hermes: "0.10",
	ios: "15.4",
	node: "16.9",
	oculus: "17.0",
	opera: "79",
	"opera-android": "66",
	opera_mobile: "66",
	quest: "17.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "15.4",
	samsung: "17.0"
},
	"es.object.is": {
	android: "4.1",
	bun: "0.1.1",
	chrome: "19",
	"chrome-android": "25",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "22",
	"firefox-android": "22",
	hermes: "0.1",
	ios: "9.0",
	node: "0.7.3",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "1.5"
},
	"es.object.is-extensible": {
	android: "44",
	bun: "0.1.1",
	chrome: "44",
	"chrome-android": "44",
	deno: "1.0",
	edge: "13",
	electron: "0.30",
	firefox: "35",
	"firefox-android": "35",
	hermes: "0.1",
	ios: "9.0",
	node: "3.0",
	oculus: "3.0",
	opera: "31",
	"opera-android": "31",
	opera_mobile: "31",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "4.0"
},
	"es.object.is-frozen": {
	android: "44",
	bun: "0.1.1",
	chrome: "44",
	"chrome-android": "44",
	deno: "1.0",
	edge: "13",
	electron: "0.30",
	firefox: "35",
	"firefox-android": "35",
	hermes: "0.1",
	ios: "9.0",
	node: "3.0",
	oculus: "3.0",
	opera: "31",
	"opera-android": "31",
	opera_mobile: "31",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "4.0"
},
	"es.object.is-sealed": {
	android: "44",
	bun: "0.1.1",
	chrome: "44",
	"chrome-android": "44",
	deno: "1.0",
	edge: "13",
	electron: "0.30",
	firefox: "35",
	"firefox-android": "35",
	hermes: "0.1",
	ios: "9.0",
	node: "3.0",
	oculus: "3.0",
	opera: "31",
	"opera-android": "31",
	opera_mobile: "31",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "4.0"
},
	"es.object.keys": {
	android: "40",
	bun: "0.1.1",
	chrome: "40",
	"chrome-android": "40",
	deno: "1.0",
	edge: "13",
	electron: "0.21",
	firefox: "35",
	"firefox-android": "35",
	hermes: "0.1",
	ios: "9.0",
	node: "1.0",
	oculus: "3.0",
	opera: "27",
	"opera-android": "27",
	opera_mobile: "27",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.4"
},
	"es.object.lookup-getter": {
	android: "62",
	bun: "0.1.1",
	chrome: "62",
	"chrome-android": "62",
	deno: "1.0",
	edge: "16",
	electron: "3.0",
	firefox: "48",
	"firefox-android": "48",
	hermes: "0.1",
	ios: "8.0",
	node: "8.10",
	oculus: "5.0",
	opera: "49",
	"opera-android": "46",
	opera_mobile: "46",
	quest: "5.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "8.0"
},
	"es.object.lookup-setter": {
	android: "62",
	bun: "0.1.1",
	chrome: "62",
	"chrome-android": "62",
	deno: "1.0",
	edge: "16",
	electron: "3.0",
	firefox: "48",
	"firefox-android": "48",
	hermes: "0.1",
	ios: "8.0",
	node: "8.10",
	oculus: "5.0",
	opera: "49",
	"opera-android": "46",
	opera_mobile: "46",
	quest: "5.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "8.0"
},
	"es.object.prevent-extensions": {
	android: "44",
	bun: "0.1.1",
	chrome: "44",
	"chrome-android": "44",
	deno: "1.0",
	edge: "13",
	electron: "0.30",
	firefox: "35",
	"firefox-android": "35",
	hermes: "0.1",
	ios: "9.0",
	node: "3.0",
	oculus: "3.0",
	opera: "31",
	"opera-android": "31",
	opera_mobile: "31",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "4.0"
},
	"es.object.proto": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "11",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "3.1",
	samsung: "1.0"
},
	"es.object.seal": {
	android: "44",
	bun: "0.1.1",
	chrome: "44",
	"chrome-android": "44",
	deno: "1.0",
	edge: "13",
	electron: "0.30",
	firefox: "35",
	"firefox-android": "35",
	hermes: "0.1",
	ios: "9.0",
	node: "3.0",
	oculus: "3.0",
	opera: "31",
	"opera-android": "31",
	opera_mobile: "31",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "4.0"
},
	"es.object.set-prototype-of": {
	android: "37",
	bun: "0.1.1",
	chrome: "34",
	"chrome-android": "34",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "31",
	"firefox-android": "31",
	hermes: "0.1",
	ie: "11",
	ios: "9.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "21",
	"opera-android": "21",
	opera_mobile: "21",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "2.0"
},
	"es.object.to-string": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "15",
	electron: "0.37",
	firefox: "51",
	"firefox-android": "51",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.object.values": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "14",
	electron: "1.4",
	firefox: "47",
	"firefox-android": "47",
	hermes: "0.1",
	ios: "10.3",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "10.1",
	samsung: "6.0"
},
	"es.parse-float": {
	android: "37",
	bun: "0.1.1",
	chrome: "35",
	"chrome-android": "35",
	deno: "1.0",
	edge: "74",
	electron: "0.20",
	firefox: "8",
	"firefox-android": "8",
	hermes: "0.1",
	ie: "8",
	ios: "8.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "22",
	"opera-android": "22",
	opera_mobile: "22",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.parse-int": {
	android: "37",
	bun: "0.1.1",
	chrome: "35",
	"chrome-android": "35",
	deno: "1.0",
	edge: "74",
	electron: "0.20",
	firefox: "21",
	"firefox-android": "21",
	hermes: "0.1",
	ie: "9",
	ios: "8.0",
	node: "0.11.13",
	oculus: "3.0",
	opera: "22",
	"opera-android": "22",
	opera_mobile: "22",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "7.1",
	samsung: "3.0"
},
	"es.promise": {
	android: "67",
	bun: "0.1.1",
	chrome: "67",
	"chrome-android": "67",
	deno: "1.24",
	edge: "79",
	electron: "4.0",
	firefox: "69",
	"firefox-android": "69",
	ios: "11.0",
	node: "10.4",
	oculus: "6.0",
	opera: "54",
	"opera-android": "48",
	opera_mobile: "48",
	quest: "6.0",
	rhino: "1.7.14",
	safari: "11.0",
	samsung: "9.0"
},
	"es.promise.all-settled": {
	android: "76",
	bun: "0.1.1",
	chrome: "76",
	"chrome-android": "76",
	deno: "1.24",
	edge: "79",
	electron: "6.0",
	firefox: "71",
	"firefox-android": "71",
	ios: "13.0",
	node: "12.9",
	oculus: "7.0",
	opera: "63",
	"opera-android": "54",
	opera_mobile: "54",
	quest: "7.0",
	rhino: "1.7.15",
	safari: "13",
	samsung: "12.0"
},
	"es.promise.any": {
	android: "85",
	bun: "0.1.1",
	chrome: "85",
	"chrome-android": "85",
	deno: "1.24",
	edge: "85",
	electron: "10.0",
	firefox: "79",
	"firefox-android": "79",
	ios: "14.0",
	node: "15.0",
	oculus: "12.0",
	opera: "71",
	"opera-android": "60",
	opera_mobile: "60",
	quest: "12.0",
	rhino: "1.8.0",
	safari: "14.0",
	samsung: "14.0"
},
	"es.promise.finally": {
	android: "67",
	bun: "0.1.1",
	chrome: "67",
	"chrome-android": "67",
	deno: "1.24",
	edge: "79",
	electron: "4.0",
	firefox: "69",
	"firefox-android": "69",
	ios: "13.2.3",
	node: "10.4",
	oculus: "6.0",
	opera: "54",
	"opera-android": "48",
	opera_mobile: "48",
	quest: "6.0",
	rhino: "1.7.14",
	safari: "13.0.3",
	samsung: "9.0"
},
	"es.promise.try": {
	android: "128",
	bun: "1.1.22",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "134",
	"firefox-android": "134",
	ios: "18.2",
	node: "23.0",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	rhino: "1.9.0",
	safari: "18.2",
	samsung: "28.0"
},
	"es.promise.with-resolvers": {
	android: "119",
	bun: "0.7.1",
	chrome: "119",
	"chrome-android": "119",
	deno: "1.38",
	edge: "119",
	electron: "28.0",
	firefox: "121",
	"firefox-android": "121",
	ios: "17.4",
	node: "22.0",
	oculus: "31.0",
	opera: "105",
	"opera-android": "79",
	opera_mobile: "79",
	quest: "31.0",
	safari: "17.4",
	samsung: "25.0"
},
	"es.array.from-async": {
	android: "121",
	bun: "1.1.2",
	chrome: "121",
	"chrome-android": "121",
	deno: "1.38",
	edge: "121",
	electron: "29.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "18.0",
	node: "22.0",
	oculus: "32.0",
	opera: "107",
	"opera-android": "81",
	opera_mobile: "81",
	quest: "32.0",
	rhino: "1.9.0",
	safari: "18.0",
	samsung: "25.0"
},
	"es.async-disposable-stack.constructor": {
	android: "136",
	bun: "1.3.0",
	chrome: "136",
	"chrome-android": "136",
	deno: "2.3.2",
	edge: "136",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "90",
	opera_mobile: "90",
	quest: "39.0",
	samsung: "29.0"
},
	"es.async-iterator.async-dispose": {
},
	"es.reflect.apply": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "15",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.construct": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "15",
	electron: "0.37",
	firefox: "44",
	"firefox-android": "44",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.define-property": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "13",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.delete-property": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "12",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.get": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "12",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.get-own-property-descriptor": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "12",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.get-prototype-of": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "12",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.has": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "12",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.is-extensible": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "12",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.own-keys": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "12",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.prevent-extensions": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "12",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.set": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "79",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.set-prototype-of": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "12",
	electron: "0.37",
	firefox: "42",
	"firefox-android": "42",
	hermes: "0.7",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.reflect.to-string-tag": {
	android: "86",
	bun: "0.1.1",
	chrome: "86",
	"chrome-android": "86",
	deno: "1.3",
	edge: "86",
	electron: "11.0",
	firefox: "82",
	"firefox-android": "82",
	hermes: "0.7",
	ios: "14.0",
	node: "15.0",
	oculus: "12.0",
	opera: "72",
	"opera-android": "61",
	opera_mobile: "61",
	quest: "12.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "14.0",
	samsung: "14.0"
},
	"es.regexp.constructor": {
	android: "64",
	bun: "0.1.1",
	chrome: "64",
	"chrome-android": "64",
	deno: "1.0",
	edge: "79",
	electron: "3.0",
	firefox: "78",
	"firefox-android": "78",
	ios: "11.3",
	node: "10.0",
	oculus: "5.0",
	opera: "51",
	"opera-android": "47",
	opera_mobile: "47",
	quest: "5.0",
	safari: "11.1",
	samsung: "9.0"
},
	"es.regexp.escape": {
	android: "136",
	bun: "1.1.22",
	chrome: "136",
	"chrome-android": "136",
	deno: "2.3.2",
	edge: "136",
	electron: "36.0",
	firefox: "134",
	"firefox-android": "134",
	ios: "18.2",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "90",
	opera_mobile: "90",
	quest: "39.0",
	safari: "18.2",
	samsung: "29.0"
},
	"es.regexp.dot-all": {
	android: "62",
	bun: "0.1.1",
	chrome: "62",
	"chrome-android": "62",
	deno: "1.0",
	edge: "79",
	electron: "3.0",
	firefox: "78",
	"firefox-android": "78",
	hermes: "0.4",
	ios: "11.3",
	node: "8.10",
	oculus: "5.0",
	opera: "49",
	"opera-android": "46",
	opera_mobile: "46",
	quest: "5.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "11.1",
	samsung: "8.0"
},
	"es.regexp.exec": {
	android: "64",
	bun: "0.1.1",
	chrome: "64",
	"chrome-android": "64",
	deno: "1.0",
	edge: "79",
	electron: "3.0",
	firefox: "78",
	"firefox-android": "78",
	hermes: "0.13",
	ios: "11.3",
	node: "10.0",
	oculus: "5.0",
	opera: "51",
	"opera-android": "47",
	opera_mobile: "47",
	quest: "5.0",
	"react-native": "0.71",
	rhino: "1.9.0",
	safari: "11.1",
	samsung: "9.0"
},
	"es.regexp.flags": {
	android: "111",
	bun: "0.1.1",
	chrome: "111",
	"chrome-android": "111",
	deno: "1.32",
	edge: "111",
	electron: "24.0",
	firefox: "78",
	"firefox-android": "78",
	hermes: "0.4",
	ios: "11.3",
	node: "20.0",
	oculus: "27.0",
	opera: "97",
	"opera-android": "75",
	opera_mobile: "75",
	quest: "27.0",
	"react-native": "0.69",
	safari: "11.1",
	samsung: "22.0"
},
	"es.regexp.sticky": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "13",
	electron: "0.37",
	firefox: "3",
	"firefox-android": "4",
	hermes: "0.3",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "10.0",
	samsung: "5.0"
},
	"es.regexp.test": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "79",
	electron: "1.2",
	firefox: "46",
	"firefox-android": "46",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.regexp.to-string": {
	android: "50",
	bun: "0.1.1",
	chrome: "50",
	"chrome-android": "50",
	deno: "1.0",
	edge: "79",
	electron: "1.1",
	firefox: "46",
	"firefox-android": "46",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "37",
	"opera-android": "37",
	opera_mobile: "37",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "10.0",
	samsung: "5.0"
},
	"es.set": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "15",
	electron: "1.2",
	firefox: "53",
	"firefox-android": "53",
	hermes: "0.13",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.73",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.set.difference.v2": {
	android: "128",
	bun: "1.2.5",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "26.0",
	node: "22.10",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	safari: "26.0",
	samsung: "28.0"
},
	"es.set.intersection.v2": {
	android: "128",
	bun: "1.1.1",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "18.0",
	node: "22.10",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	safari: "18.0",
	samsung: "28.0"
},
	"es.set.is-disjoint-from.v2": {
	android: "128",
	bun: "1.1.1",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "18.0",
	node: "22.10",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	safari: "18.0",
	samsung: "28.0"
},
	"es.set.is-subset-of.v2": {
	android: "128",
	bun: "1.1.1",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "18.0",
	node: "22.10",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	safari: "18.0",
	samsung: "28.0"
},
	"es.set.is-superset-of.v2": {
	android: "128",
	bun: "1.1.1",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "18.0",
	node: "22.10",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	safari: "18.0",
	samsung: "28.0"
},
	"es.set.symmetric-difference.v2": {
	android: "123",
	bun: "1.2.5",
	chrome: "123",
	"chrome-android": "123",
	deno: "1.41.3",
	edge: "123",
	electron: "30.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "26.0",
	node: "22.0",
	oculus: "33.0",
	opera: "109",
	"opera-android": "82",
	opera_mobile: "82",
	quest: "33.0",
	safari: "26.0",
	samsung: "27.0"
},
	"es.set.union.v2": {
	android: "123",
	bun: "1.2.5",
	chrome: "123",
	"chrome-android": "123",
	deno: "1.41.3",
	edge: "123",
	electron: "30.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "26.0",
	node: "22.0",
	oculus: "33.0",
	opera: "109",
	"opera-android": "82",
	opera_mobile: "82",
	quest: "33.0",
	safari: "26.0",
	samsung: "27.0"
},
	"es.string.at-alternative": {
	android: "92",
	bun: "0.1.1",
	chrome: "92",
	"chrome-android": "92",
	deno: "1.12",
	edge: "92",
	electron: "14.0",
	firefox: "90",
	"firefox-android": "90",
	hermes: "0.13",
	ios: "15.4",
	node: "16.6",
	oculus: "17.0",
	opera: "78",
	"opera-android": "65",
	opera_mobile: "65",
	quest: "17.0",
	"react-native": "0.71",
	safari: "15.4",
	samsung: "16.0"
},
	"es.string.code-point-at": {
	android: "41",
	bun: "0.1.1",
	chrome: "41",
	"chrome-android": "41",
	deno: "1.0",
	edge: "13",
	electron: "0.21",
	firefox: "29",
	"firefox-android": "29",
	hermes: "0.1",
	ios: "9.0",
	node: "1.0",
	oculus: "3.0",
	opera: "28",
	"opera-android": "28",
	opera_mobile: "28",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.4"
},
	"es.string.ends-with": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "79",
	electron: "1.2",
	firefox: "40",
	"firefox-android": "40",
	hermes: "0.1",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "10.0",
	samsung: "5.0"
},
	"es.string.from-code-point": {
	android: "41",
	bun: "0.1.1",
	chrome: "41",
	"chrome-android": "41",
	deno: "1.0",
	edge: "13",
	electron: "0.21",
	firefox: "29",
	"firefox-android": "29",
	hermes: "0.1",
	ios: "9.0",
	node: "1.0",
	oculus: "3.0",
	opera: "28",
	"opera-android": "28",
	opera_mobile: "28",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.4"
},
	"es.string.includes": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "79",
	electron: "1.2",
	firefox: "40",
	"firefox-android": "40",
	hermes: "0.1",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "10.0",
	samsung: "5.0"
},
	"es.string.is-well-formed": {
	android: "111",
	bun: "0.4.0",
	chrome: "111",
	"chrome-android": "111",
	deno: "1.32",
	edge: "111",
	electron: "24.0",
	firefox: "119",
	"firefox-android": "119",
	ios: "16.4",
	node: "20.0",
	oculus: "27.0",
	opera: "97",
	"opera-android": "75",
	opera_mobile: "75",
	quest: "27.0",
	rhino: "1.8.0",
	safari: "16.4",
	samsung: "22.0"
},
	"es.string.iterator": {
	android: "41",
	bun: "0.1.1",
	chrome: "41",
	"chrome-android": "41",
	deno: "1.0",
	edge: "13",
	electron: "0.21",
	firefox: "36",
	"firefox-android": "36",
	hermes: "0.1",
	ios: "9.0",
	node: "1.0",
	oculus: "3.0",
	opera: "28",
	"opera-android": "28",
	opera_mobile: "28",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.4"
},
	"es.string.match": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "79",
	electron: "1.2",
	firefox: "49",
	"firefox-android": "49",
	hermes: "0.1",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.string.match-all": {
	android: "80",
	bun: "0.1.1",
	chrome: "80",
	"chrome-android": "80",
	deno: "1.0",
	edge: "80",
	electron: "8.0",
	firefox: "73",
	"firefox-android": "73",
	hermes: "0.6",
	ios: "13.4",
	node: "14.0",
	oculus: "9.0",
	opera: "67",
	"opera-android": "57",
	opera_mobile: "57",
	quest: "9.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "13.1",
	samsung: "13.0"
},
	"es.string.pad-end": {
	android: "57",
	bun: "0.1.1",
	chrome: "57",
	"chrome-android": "57",
	deno: "1.0",
	edge: "15",
	electron: "1.7",
	firefox: "48",
	"firefox-android": "48",
	hermes: "0.1",
	ios: "11.0",
	node: "8.0",
	oculus: "3.0",
	opera: "44",
	"opera-android": "43",
	opera_mobile: "43",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "11.0",
	samsung: "7.0"
},
	"es.string.pad-start": {
	android: "57",
	bun: "0.1.1",
	chrome: "57",
	"chrome-android": "57",
	deno: "1.0",
	edge: "15",
	electron: "1.7",
	firefox: "48",
	"firefox-android": "48",
	hermes: "0.1",
	ios: "11.0",
	node: "8.0",
	oculus: "3.0",
	opera: "44",
	"opera-android": "43",
	opera_mobile: "43",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "11.0",
	samsung: "7.0"
},
	"es.string.raw": {
	android: "41",
	bun: "0.1.1",
	chrome: "41",
	"chrome-android": "41",
	deno: "1.0",
	edge: "13",
	electron: "0.21",
	firefox: "34",
	"firefox-android": "34",
	hermes: "0.1",
	ios: "9.0",
	node: "1.0",
	oculus: "3.0",
	opera: "28",
	"opera-android": "28",
	opera_mobile: "28",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "9.0",
	samsung: "3.4"
},
	"es.string.repeat": {
	android: "41",
	bun: "0.1.1",
	chrome: "41",
	"chrome-android": "41",
	deno: "1.0",
	edge: "13",
	electron: "0.21",
	firefox: "24",
	"firefox-android": "24",
	hermes: "0.1",
	ios: "9.0",
	node: "1.0",
	oculus: "3.0",
	opera: "28",
	"opera-android": "28",
	opera_mobile: "28",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "9.0",
	samsung: "3.4"
},
	"es.string.replace": {
	android: "64",
	bun: "0.1.1",
	chrome: "64",
	"chrome-android": "64",
	deno: "1.0",
	edge: "79",
	electron: "3.0",
	firefox: "78",
	"firefox-android": "78",
	hermes: "0.13",
	ios: "14.0",
	node: "10.0",
	oculus: "5.0",
	opera: "51",
	"opera-android": "47",
	opera_mobile: "47",
	quest: "5.0",
	"react-native": "0.71",
	rhino: "1.9.0",
	safari: "14.0",
	samsung: "9.0"
},
	"es.string.replace-all": {
	android: "85",
	bun: "0.1.1",
	chrome: "85",
	"chrome-android": "85",
	deno: "1.2",
	edge: "85",
	electron: "10.0",
	firefox: "77",
	"firefox-android": "77",
	hermes: "0.7",
	ios: "13.4",
	node: "15.0",
	oculus: "12.0",
	opera: "71",
	"opera-android": "60",
	opera_mobile: "60",
	quest: "12.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "13.1",
	samsung: "14.0"
},
	"es.string.search": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "79",
	electron: "1.2",
	firefox: "49",
	"firefox-android": "49",
	hermes: "0.1",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.string.split": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "79",
	electron: "1.4",
	firefox: "49",
	"firefox-android": "49",
	ios: "10.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "6.0"
},
	"es.string.starts-with": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "79",
	electron: "1.2",
	firefox: "40",
	"firefox-android": "40",
	hermes: "0.1",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "10.0",
	samsung: "5.0"
},
	"es.string.substr": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "1",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "1",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "9",
	ios: "1.0",
	node: "0.0.3",
	oculus: "3.0",
	opera: "4",
	"opera-android": "4",
	opera_mobile: "4",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "1",
	samsung: "1.0"
},
	"es.string.to-well-formed": {
	android: "111",
	bun: "0.5.7",
	chrome: "111",
	"chrome-android": "111",
	deno: "1.32",
	edge: "111",
	electron: "24.0",
	firefox: "119",
	"firefox-android": "119",
	ios: "16.4",
	node: "20.0",
	oculus: "27.0",
	opera: "97",
	"opera-android": "75",
	opera_mobile: "75",
	quest: "27.0",
	rhino: "1.8.0",
	safari: "16.4",
	samsung: "22.0"
},
	"es.string.trim": {
	android: "59",
	bun: "0.1.1",
	chrome: "59",
	"chrome-android": "59",
	deno: "1.0",
	edge: "15",
	electron: "1.8",
	firefox: "52",
	"firefox-android": "52",
	hermes: "0.1",
	ios: "12.2",
	node: "8.3",
	oculus: "4.0",
	opera: "46",
	"opera-android": "43",
	opera_mobile: "43",
	quest: "4.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "12.1",
	samsung: "7.0"
},
	"es.string.trim-end": {
	android: "66",
	bun: "0.1.1",
	chrome: "66",
	"chrome-android": "66",
	deno: "1.0",
	edge: "79",
	electron: "3.0",
	firefox: "61",
	"firefox-android": "61",
	hermes: "0.3",
	ios: "12.2",
	node: "10.0",
	oculus: "5.0",
	opera: "53",
	"opera-android": "47",
	opera_mobile: "47",
	quest: "5.0",
	"react-native": "0.69",
	safari: "12.1",
	samsung: "9.0"
},
	"es.string.trim-start": {
	android: "66",
	bun: "0.1.1",
	chrome: "66",
	"chrome-android": "66",
	deno: "1.0",
	edge: "79",
	electron: "3.0",
	firefox: "61",
	"firefox-android": "61",
	hermes: "0.3",
	ios: "12.0",
	node: "10.0",
	oculus: "5.0",
	opera: "53",
	"opera-android": "47",
	opera_mobile: "47",
	quest: "5.0",
	"react-native": "0.69",
	safari: "12.0",
	samsung: "9.0"
},
	"es.string.anchor": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "17",
	"firefox-android": "17",
	ios: "6.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	phantom: "2.0",
	quest: "3.0",
	rhino: "1.7.14",
	safari: "6.0",
	samsung: "1.0"
},
	"es.string.big": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "3.1",
	samsung: "1.0"
},
	"es.string.blink": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "3.1",
	samsung: "1.0"
},
	"es.string.bold": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "3.1",
	samsung: "1.0"
},
	"es.string.fixed": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "3.1",
	samsung: "1.0"
},
	"es.string.fontcolor": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "17",
	"firefox-android": "17",
	ios: "6.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	phantom: "2.0",
	quest: "3.0",
	rhino: "1.7.14",
	safari: "6.0",
	samsung: "1.0"
},
	"es.string.fontsize": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "17",
	"firefox-android": "17",
	ios: "6.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	phantom: "2.0",
	quest: "3.0",
	rhino: "1.7.14",
	safari: "6.0",
	samsung: "1.0"
},
	"es.string.italics": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "3.1",
	samsung: "1.0"
},
	"es.string.link": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "17",
	"firefox-android": "17",
	ios: "6.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	phantom: "2.0",
	quest: "3.0",
	rhino: "1.7.14",
	safari: "6.0",
	samsung: "1.0"
},
	"es.string.small": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "3.1",
	samsung: "1.0"
},
	"es.string.strike": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "3.1",
	samsung: "1.0"
},
	"es.string.sub": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "3.1",
	samsung: "1.0"
},
	"es.string.sup": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "5",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "2",
	"firefox-android": "4",
	ios: "2.0",
	node: "0.1.27",
	oculus: "3.0",
	opera: "10.50",
	"opera-android": "10.50",
	opera_mobile: "10.50",
	phantom: "1.9",
	quest: "3.0",
	rhino: "1.7.13",
	safari: "3.1",
	samsung: "1.0"
},
	"es.typed-array.float32-array": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.float64-array": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.int8-array": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.int16-array": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.int32-array": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.uint8-array": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.uint8-clamped-array": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.uint16-array": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.uint32-array": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.at": {
	android: "92",
	bun: "0.1.1",
	chrome: "92",
	"chrome-android": "92",
	deno: "1.12",
	edge: "92",
	electron: "14.0",
	firefox: "90",
	"firefox-android": "90",
	hermes: "0.13",
	ios: "15.4",
	node: "16.6",
	oculus: "17.0",
	opera: "78",
	"opera-android": "65",
	opera_mobile: "65",
	quest: "17.0",
	"react-native": "0.71",
	rhino: "1.7.15",
	safari: "15.4",
	samsung: "16.0"
},
	"es.typed-array.copy-within": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "34",
	"firefox-android": "34",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.every": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.fill": {
	android: "58",
	bun: "0.1.1",
	chrome: "58",
	"chrome-android": "58",
	deno: "1.0",
	edge: "79",
	electron: "1.7",
	firefox: "55",
	"firefox-android": "55",
	hermes: "0.1",
	ios: "14.5",
	node: "8.0",
	oculus: "4.0",
	opera: "45",
	"opera-android": "43",
	opera_mobile: "43",
	quest: "4.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "14.1",
	samsung: "7.0"
},
	"es.typed-array.filter": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "38",
	"firefox-android": "38",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.find": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.find-index": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.find-last": {
	android: "97",
	bun: "0.1.1",
	chrome: "97",
	"chrome-android": "97",
	deno: "1.16",
	edge: "97",
	electron: "17.0",
	firefox: "104",
	"firefox-android": "104",
	hermes: "0.11",
	ios: "15.4",
	node: "18.0",
	oculus: "20.0",
	opera: "83",
	"opera-android": "68",
	opera_mobile: "68",
	quest: "20.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "15.4",
	samsung: "18.0"
},
	"es.typed-array.find-last-index": {
	android: "97",
	bun: "0.1.1",
	chrome: "97",
	"chrome-android": "97",
	deno: "1.16",
	edge: "97",
	electron: "17.0",
	firefox: "104",
	"firefox-android": "104",
	hermes: "0.11",
	ios: "15.4",
	node: "18.0",
	oculus: "20.0",
	opera: "83",
	"opera-android": "68",
	opera_mobile: "68",
	quest: "20.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "15.4",
	samsung: "18.0"
},
	"es.typed-array.for-each": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "38",
	"firefox-android": "38",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.from": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.includes": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.0",
	edge: "14",
	electron: "0.37",
	firefox: "43",
	"firefox-android": "43",
	hermes: "0.1",
	ios: "10.0",
	node: "6.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.index-of": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.iterator": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "13",
	electron: "1.2",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.join": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.last-index-of": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.map": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "38",
	"firefox-android": "38",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.of": {
	android: "54",
	bun: "0.1.1",
	chrome: "54",
	"chrome-android": "54",
	deno: "1.0",
	edge: "15",
	electron: "1.4",
	firefox: "55",
	"firefox-android": "55",
	ios: "14.0",
	node: "7.0",
	oculus: "3.0",
	opera: "41",
	"opera-android": "41",
	opera_mobile: "41",
	quest: "3.0",
	safari: "14.0",
	samsung: "6.0"
},
	"es.typed-array.reduce": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.reduce-right": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.reverse": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.set": {
	android: "95",
	bun: "0.1.1",
	chrome: "95",
	"chrome-android": "95",
	deno: "1.15",
	edge: "95",
	electron: "16.0",
	firefox: "54",
	"firefox-android": "54",
	hermes: "0.1",
	ios: "14.5",
	node: "17.0",
	oculus: "18.0",
	opera: "81",
	"opera-android": "67",
	opera_mobile: "67",
	quest: "18.0",
	"react-native": "0.69",
	safari: "14.1",
	samsung: "17.0"
},
	"es.typed-array.slice": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "38",
	"firefox-android": "38",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.some": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "13",
	electron: "0.31",
	firefox: "37",
	"firefox-android": "37",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.sort": {
	android: "74",
	bun: "0.1.1",
	chrome: "74",
	"chrome-android": "74",
	deno: "1.0",
	edge: "79",
	electron: "6.0",
	firefox: "67",
	"firefox-android": "67",
	hermes: "0.10",
	ios: "14.5",
	node: "12.0",
	oculus: "6.0",
	opera: "61",
	"opera-android": "53",
	opera_mobile: "53",
	quest: "6.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "14.1",
	samsung: "11.0"
},
	"es.typed-array.subarray": {
	android: "4.4",
	bun: "0.1.1",
	chrome: "26",
	"chrome-android": "26",
	deno: "1.0",
	edge: "13",
	electron: "0.20",
	firefox: "15",
	"firefox-android": "15",
	hermes: "0.1",
	ios: "8.0",
	node: "0.11.0",
	oculus: "3.0",
	opera: "15",
	"opera-android": "15",
	opera_mobile: "15",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "7.1",
	samsung: "1.5"
},
	"es.typed-array.to-locale-string": {
	android: "45",
	bun: "0.1.1",
	chrome: "45",
	"chrome-android": "45",
	deno: "1.0",
	edge: "79",
	electron: "0.31",
	firefox: "51",
	"firefox-android": "51",
	hermes: "0.1",
	ios: "10.0",
	node: "4.0",
	oculus: "3.0",
	opera: "32",
	"opera-android": "32",
	opera_mobile: "32",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.9.0",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.to-reversed": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"es.typed-array.to-sorted": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"es.typed-array.to-string": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "13",
	electron: "1.2",
	firefox: "51",
	"firefox-android": "51",
	hermes: "0.1",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.69",
	safari: "10.0",
	samsung: "5.0"
},
	"es.typed-array.with": {
	android: "110",
	bun: "1.2.18",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "26.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	rhino: "1.8.0",
	safari: "26.0",
	samsung: "21.0"
},
	"es.uint8-array.from-base64": {
	android: "140",
	bun: "1.2.20",
	chrome: "140",
	"chrome-android": "140",
	deno: "2.5",
	edge: "140",
	electron: "38.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "26.2",
	node: "25.0",
	oculus: "41.0",
	opera: "124",
	"opera-android": "92",
	opera_mobile: "92",
	quest: "41.0",
	safari: "26.2"
},
	"es.uint8-array.from-hex": {
	android: "140",
	bun: "1.1.22",
	chrome: "140",
	"chrome-android": "140",
	deno: "2.5",
	edge: "140",
	electron: "38.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "18.2",
	node: "25.0",
	oculus: "41.0",
	opera: "124",
	"opera-android": "92",
	opera_mobile: "92",
	quest: "41.0",
	safari: "18.2"
},
	"es.uint8-array.set-from-base64": {
	android: "140",
	bun: "1.2.20",
	chrome: "140",
	"chrome-android": "140",
	deno: "2.5",
	edge: "140",
	electron: "38.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "26.2",
	node: "25.0",
	oculus: "41.0",
	opera: "124",
	"opera-android": "92",
	opera_mobile: "92",
	quest: "41.0",
	safari: "26.2"
},
	"es.uint8-array.set-from-hex": {
	android: "144",
	bun: "1.1.22",
	chrome: "144",
	"chrome-android": "144",
	edge: "144",
	electron: "40.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "18.2",
	opera: "128",
	safari: "18.2"
},
	"es.uint8-array.to-base64": {
	android: "140",
	bun: "1.1.22",
	chrome: "140",
	"chrome-android": "140",
	deno: "2.5",
	edge: "140",
	electron: "38.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "18.2",
	node: "25.0",
	oculus: "41.0",
	opera: "124",
	"opera-android": "92",
	opera_mobile: "92",
	quest: "41.0",
	safari: "18.2"
},
	"es.uint8-array.to-hex": {
	android: "140",
	bun: "1.1.22",
	chrome: "140",
	"chrome-android": "140",
	deno: "2.5",
	edge: "140",
	electron: "38.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "18.2",
	node: "25.0",
	oculus: "41.0",
	opera: "124",
	"opera-android": "92",
	opera_mobile: "92",
	quest: "41.0",
	safari: "18.2"
},
	"es.unescape": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "1",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "1",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "3",
	ios: "1.0",
	node: "0.0.3",
	oculus: "3.0",
	opera: "3",
	"opera-android": "3",
	opera_mobile: "3",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "1",
	samsung: "1.0"
},
	"es.weak-map": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "79",
	electron: "1.2",
	firefox: "53",
	"firefox-android": "53",
	hermes: "0.13",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.73",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"es.weak-map.get-or-insert": {
	android: "145",
	bun: "1.2.20",
	chrome: "145",
	"chrome-android": "145",
	edge: "145",
	electron: "41.0",
	firefox: "144",
	"firefox-android": "144",
	ios: "26.2",
	opera: "129",
	safari: "26.2"
},
	"es.weak-map.get-or-insert-computed": {
	android: "145",
	bun: "1.2.20",
	chrome: "145",
	"chrome-android": "145",
	edge: "145",
	electron: "41.0",
	firefox: "144",
	"firefox-android": "144",
	ios: "26.2",
	opera: "129",
	safari: "26.2"
},
	"es.weak-set": {
	android: "51",
	bun: "0.1.1",
	chrome: "51",
	"chrome-android": "51",
	deno: "1.0",
	edge: "15",
	electron: "1.2",
	firefox: "53",
	"firefox-android": "53",
	hermes: "0.13",
	ios: "10.0",
	node: "6.5",
	oculus: "3.0",
	opera: "38",
	"opera-android": "38",
	opera_mobile: "38",
	quest: "3.0",
	"react-native": "0.73",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "5.0"
},
	"esnext.aggregate-error": {
	android: "85",
	bun: "0.1.1",
	chrome: "85",
	"chrome-android": "85",
	deno: "1.2",
	edge: "85",
	electron: "10.0",
	firefox: "79",
	"firefox-android": "79",
	hermes: "0.13",
	ios: "14.0",
	node: "15.0",
	oculus: "12.0",
	opera: "71",
	"opera-android": "60",
	opera_mobile: "60",
	quest: "12.0",
	"react-native": "0.72",
	rhino: "1.8.0",
	safari: "14.0",
	samsung: "14.0"
},
	"esnext.suppressed-error.constructor": {
	android: "136",
	bun: "1.2.15",
	chrome: "136",
	"chrome-android": "136",
	deno: "2.2.10",
	edge: "136",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "90",
	opera_mobile: "90",
	quest: "39.0",
	samsung: "29.0"
},
	"esnext.array.from-async": {
	android: "121",
	bun: "1.1.2",
	chrome: "121",
	"chrome-android": "121",
	deno: "1.38",
	edge: "121",
	electron: "29.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "18.0",
	node: "22.0",
	oculus: "32.0",
	opera: "107",
	"opera-android": "81",
	opera_mobile: "81",
	quest: "32.0",
	rhino: "1.9.0",
	safari: "18.0",
	samsung: "25.0"
},
	"esnext.array.at": {
	android: "92",
	bun: "0.1.1",
	chrome: "92",
	"chrome-android": "92",
	deno: "1.12",
	edge: "92",
	electron: "14.0",
	firefox: "90",
	"firefox-android": "90",
	hermes: "0.13",
	ios: "15.4",
	node: "16.6",
	oculus: "17.0",
	opera: "78",
	"opera-android": "65",
	opera_mobile: "65",
	quest: "17.0",
	"react-native": "0.71",
	rhino: "1.7.15",
	safari: "15.4",
	samsung: "16.0"
},
	"esnext.array.filter-out": {
},
	"esnext.array.filter-reject": {
},
	"esnext.array.find-last": {
	android: "97",
	bun: "0.1.1",
	chrome: "97",
	"chrome-android": "97",
	deno: "1.16",
	edge: "97",
	electron: "17.0",
	firefox: "104",
	"firefox-android": "104",
	hermes: "0.11",
	ios: "15.4",
	node: "18.0",
	oculus: "20.0",
	opera: "83",
	"opera-android": "68",
	opera_mobile: "68",
	quest: "20.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "15.4",
	samsung: "18.0"
},
	"esnext.array.find-last-index": {
	android: "97",
	bun: "0.1.1",
	chrome: "97",
	"chrome-android": "97",
	deno: "1.16",
	edge: "97",
	electron: "17.0",
	firefox: "104",
	"firefox-android": "104",
	hermes: "0.11",
	ios: "15.4",
	node: "18.0",
	oculus: "20.0",
	opera: "83",
	"opera-android": "68",
	opera_mobile: "68",
	quest: "20.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "15.4",
	samsung: "18.0"
},
	"esnext.array.group": {
},
	"esnext.array.group-by": {
},
	"esnext.array.group-by-to-map": {
},
	"esnext.array.group-to-map": {
},
	"esnext.array.is-template-object": {
},
	"esnext.array.last-index": {
},
	"esnext.array.last-item": {
},
	"esnext.array.to-reversed": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	hermes: "0.13",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	"react-native": "0.74",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"esnext.array.to-sorted": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"esnext.array.to-spliced": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	hermes: "0.13",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	"react-native": "0.74",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"esnext.array.unique-by": {
},
	"esnext.array.with": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "140",
	"firefox-android": "140",
	hermes: "0.13",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	"react-native": "0.74",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"esnext.array-buffer.detached": {
	android: "114",
	bun: "1.0.19",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "122",
	"firefox-android": "122",
	ios: "17.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	rhino: "1.9.0",
	safari: "17.4",
	samsung: "23.0"
},
	"esnext.array-buffer.transfer": {
	android: "114",
	bun: "1.0.19",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "122",
	"firefox-android": "122",
	ios: "17.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	rhino: "1.9.0",
	safari: "17.4",
	samsung: "23.0"
},
	"esnext.array-buffer.transfer-to-fixed-length": {
	android: "114",
	bun: "1.0.19",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "122",
	"firefox-android": "122",
	ios: "17.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	rhino: "1.9.0",
	safari: "17.4",
	samsung: "23.0"
},
	"esnext.async-disposable-stack.constructor": {
	android: "136",
	bun: "1.3.0",
	chrome: "136",
	"chrome-android": "136",
	deno: "2.3.2",
	edge: "136",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "90",
	opera_mobile: "90",
	quest: "39.0",
	samsung: "29.0"
},
	"esnext.async-iterator.constructor": {
},
	"esnext.async-iterator.as-indexed-pairs": {
},
	"esnext.async-iterator.async-dispose": {
},
	"esnext.async-iterator.drop": {
},
	"esnext.async-iterator.every": {
},
	"esnext.async-iterator.filter": {
},
	"esnext.async-iterator.find": {
},
	"esnext.async-iterator.flat-map": {
},
	"esnext.async-iterator.for-each": {
},
	"esnext.async-iterator.from": {
},
	"esnext.async-iterator.indexed": {
},
	"esnext.async-iterator.map": {
},
	"esnext.async-iterator.reduce": {
},
	"esnext.async-iterator.some": {
},
	"esnext.async-iterator.take": {
},
	"esnext.async-iterator.to-array": {
},
	"esnext.bigint.range": {
},
	"esnext.composite-key": {
},
	"esnext.composite-symbol": {
},
	"esnext.data-view.get-float16": {
	android: "135",
	bun: "1.1.23",
	chrome: "135",
	"chrome-android": "135",
	deno: "1.43",
	edge: "135",
	electron: "36.0",
	firefox: "129",
	"firefox-android": "129",
	ios: "18.2",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "18.2",
	samsung: "29.0"
},
	"esnext.data-view.get-uint8-clamped": {
},
	"esnext.data-view.set-float16": {
	android: "135",
	bun: "1.1.23",
	chrome: "135",
	"chrome-android": "135",
	deno: "1.43",
	edge: "135",
	electron: "36.0",
	firefox: "129",
	"firefox-android": "129",
	ios: "18.2",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "18.2",
	samsung: "29.0"
},
	"esnext.data-view.set-uint8-clamped": {
},
	"esnext.disposable-stack.constructor": {
	android: "136",
	bun: "1.3.0",
	chrome: "136",
	"chrome-android": "136",
	deno: "2.2.10",
	edge: "136",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "90",
	opera_mobile: "90",
	quest: "39.0",
	samsung: "29.0"
},
	"esnext.error.is-error": {
	android: "134",
	chrome: "134",
	"chrome-android": "134",
	deno: "2.2",
	edge: "134",
	electron: "35.0",
	firefox: "138",
	"firefox-android": "138",
	node: "24.3",
	oculus: "38.0",
	opera: "119",
	"opera-android": "88",
	opera_mobile: "88",
	quest: "38.0",
	rhino: "1.9.0",
	samsung: "29.0"
},
	"esnext.function.demethodize": {
},
	"esnext.function.is-callable": {
},
	"esnext.function.is-constructor": {
},
	"esnext.function.metadata": {
},
	"esnext.function.un-this": {
},
	"esnext.global-this": {
	android: "71",
	bun: "0.1.1",
	chrome: "71",
	"chrome-android": "71",
	deno: "1.0",
	edge: "79",
	electron: "5.0",
	firefox: "65",
	"firefox-android": "65",
	hermes: "0.2",
	ios: "12.2",
	node: "12.0",
	oculus: "6.0",
	opera: "58",
	"opera-android": "50",
	opera_mobile: "50",
	quest: "6.0",
	"react-native": "0.69",
	rhino: "1.7.14",
	safari: "12.1",
	samsung: "10.0"
},
	"esnext.iterator.constructor": {
	android: "122",
	bun: "1.1.31",
	chrome: "122",
	"chrome-android": "122",
	deno: "1.38.1",
	edge: "122",
	electron: "29.0",
	firefox: "131",
	"firefox-android": "131",
	ios: "18.4",
	node: "22.0",
	oculus: "32.0",
	opera: "108",
	"opera-android": "81",
	opera_mobile: "81",
	quest: "32.0",
	safari: "18.4",
	samsung: "26.0"
},
	"esnext.iterator.as-indexed-pairs": {
},
	"esnext.iterator.chunks": {
},
	"esnext.iterator.concat": {
	bun: "1.3.7",
	firefox: "147",
	"firefox-android": "147"
},
	"esnext.iterator.dispose": {
	android: "136",
	bun: "1.3.0",
	chrome: "136",
	"chrome-android": "136",
	deno: "2.2.10",
	edge: "136",
	electron: "36.0",
	firefox: "135",
	"firefox-android": "135",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "90",
	opera_mobile: "90",
	quest: "39.0",
	samsung: "29.0"
},
	"esnext.iterator.drop": {
	android: "135",
	bun: "1.2.11",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"esnext.iterator.every": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"esnext.iterator.filter": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"esnext.iterator.find": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"esnext.iterator.flat-map": {
	android: "135",
	bun: "1.2.21",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.2",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.2",
	samsung: "29.0"
},
	"esnext.iterator.for-each": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"esnext.iterator.from": {
	android: "122",
	bun: "1.2.5",
	chrome: "122",
	"chrome-android": "122",
	deno: "1.38.1",
	edge: "122",
	electron: "29.0",
	firefox: "131",
	"firefox-android": "131",
	ios: "26.0",
	node: "22.0",
	oculus: "32.0",
	opera: "108",
	"opera-android": "81",
	opera_mobile: "81",
	quest: "32.0",
	safari: "26.0",
	samsung: "26.0"
},
	"esnext.iterator.indexed": {
},
	"esnext.iterator.map": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"esnext.iterator.range": {
},
	"esnext.iterator.reduce": {
	android: "135",
	bun: "1.2.11",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"esnext.iterator.sliding": {
},
	"esnext.iterator.some": {
	android: "135",
	bun: "1.2.4",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"esnext.iterator.take": {
	android: "135",
	bun: "1.2.11",
	chrome: "135",
	"chrome-android": "135",
	deno: "2.2.5",
	edge: "135",
	electron: "36.0",
	firefox: "141",
	"firefox-android": "141",
	ios: "26.0",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	safari: "26.0",
	samsung: "29.0"
},
	"esnext.iterator.to-array": {
	android: "122",
	bun: "1.1.31",
	chrome: "122",
	"chrome-android": "122",
	deno: "1.38.1",
	edge: "122",
	electron: "29.0",
	firefox: "131",
	"firefox-android": "131",
	ios: "18.4",
	node: "22.0",
	oculus: "32.0",
	opera: "108",
	"opera-android": "81",
	opera_mobile: "81",
	quest: "32.0",
	safari: "18.4",
	samsung: "26.0"
},
	"esnext.iterator.to-async": {
},
	"esnext.iterator.windows": {
},
	"esnext.iterator.zip": {
	firefox: "148",
	"firefox-android": "148"
},
	"esnext.iterator.zip-keyed": {
	firefox: "148",
	"firefox-android": "148"
},
	"esnext.json.is-raw-json": {
	android: "114",
	bun: "1.1.43",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "135",
	"firefox-android": "135",
	ios: "18.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	safari: "18.4",
	samsung: "23.0"
},
	"esnext.json.parse": {
	android: "114",
	bun: "1.1.43",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "135",
	"firefox-android": "135",
	ios: "18.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	safari: "18.4",
	samsung: "23.0"
},
	"esnext.json.raw-json": {
	android: "114",
	bun: "1.1.43",
	chrome: "114",
	"chrome-android": "114",
	deno: "1.33",
	edge: "114",
	electron: "25.0",
	firefox: "135",
	"firefox-android": "135",
	ios: "18.4",
	node: "21.0",
	oculus: "28.0",
	opera: "100",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	safari: "18.4",
	samsung: "23.0"
},
	"esnext.map.delete-all": {
},
	"esnext.map.emplace": {
},
	"esnext.map.every": {
},
	"esnext.map.filter": {
},
	"esnext.map.find": {
},
	"esnext.map.find-key": {
},
	"esnext.map.from": {
},
	"esnext.map.get-or-insert": {
	android: "145",
	bun: "1.2.20",
	chrome: "145",
	"chrome-android": "145",
	edge: "145",
	electron: "41.0",
	firefox: "144",
	"firefox-android": "144",
	ios: "26.2",
	opera: "129",
	safari: "26.2"
},
	"esnext.map.get-or-insert-computed": {
	android: "145",
	bun: "1.2.20",
	chrome: "145",
	"chrome-android": "145",
	edge: "145",
	electron: "41.0",
	firefox: "144",
	"firefox-android": "144",
	ios: "26.2",
	opera: "129",
	safari: "26.2"
},
	"esnext.map.group-by": {
	android: "117",
	bun: "1.1.2",
	chrome: "117",
	"chrome-android": "117",
	deno: "1.37",
	edge: "117",
	electron: "27.0",
	firefox: "119",
	"firefox-android": "119",
	ios: "18.0",
	node: "21.0",
	oculus: "30.0",
	opera: "103",
	"opera-android": "78",
	opera_mobile: "78",
	quest: "30.0",
	rhino: "1.8.0",
	safari: "18.0",
	samsung: "24.0"
},
	"esnext.map.includes": {
},
	"esnext.map.key-by": {
},
	"esnext.map.key-of": {
},
	"esnext.map.map-keys": {
},
	"esnext.map.map-values": {
},
	"esnext.map.merge": {
},
	"esnext.map.of": {
},
	"esnext.map.reduce": {
},
	"esnext.map.some": {
},
	"esnext.map.update": {
},
	"esnext.map.update-or-insert": {
},
	"esnext.map.upsert": {
},
	"esnext.math.clamp": {
},
	"esnext.math.deg-per-rad": {
},
	"esnext.math.degrees": {
},
	"esnext.math.fscale": {
},
	"esnext.math.f16round": {
	android: "135",
	bun: "1.1.23",
	chrome: "135",
	"chrome-android": "135",
	deno: "1.43",
	edge: "135",
	electron: "36.0",
	firefox: "129",
	"firefox-android": "129",
	ios: "18.2",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "89",
	opera_mobile: "89",
	quest: "39.0",
	rhino: "1.9.0",
	safari: "18.2",
	samsung: "29.0"
},
	"esnext.math.iaddh": {
},
	"esnext.math.imulh": {
},
	"esnext.math.isubh": {
},
	"esnext.math.rad-per-deg": {
},
	"esnext.math.radians": {
},
	"esnext.math.scale": {
},
	"esnext.math.seeded-prng": {
},
	"esnext.math.signbit": {
},
	"esnext.math.sum-precise": {
	bun: "1.2.18",
	firefox: "137",
	"firefox-android": "137",
	ios: "26.2",
	safari: "26.2"
},
	"esnext.math.umulh": {
},
	"esnext.number.clamp": {
},
	"esnext.number.from-string": {
},
	"esnext.number.range": {
},
	"esnext.object.has-own": {
	android: "93",
	bun: "0.1.1",
	chrome: "93",
	"chrome-android": "93",
	deno: "1.13",
	edge: "93",
	electron: "14.0",
	firefox: "92",
	"firefox-android": "92",
	hermes: "0.10",
	ios: "15.4",
	node: "16.9",
	oculus: "17.0",
	opera: "79",
	"opera-android": "66",
	opera_mobile: "66",
	quest: "17.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "15.4",
	samsung: "17.0"
},
	"esnext.object.iterate-entries": {
},
	"esnext.object.iterate-keys": {
},
	"esnext.object.iterate-values": {
},
	"esnext.object.group-by": {
	android: "117",
	bun: "1.1.2",
	chrome: "117",
	"chrome-android": "117",
	deno: "1.37",
	edge: "117",
	electron: "27.0",
	firefox: "119",
	"firefox-android": "119",
	ios: "18.0",
	node: "21.0",
	oculus: "30.0",
	opera: "103",
	"opera-android": "78",
	opera_mobile: "78",
	quest: "30.0",
	rhino: "1.8.0",
	safari: "18.0",
	samsung: "24.0"
},
	"esnext.observable": {
},
	"esnext.promise.all-settled": {
	android: "76",
	bun: "0.1.1",
	chrome: "76",
	"chrome-android": "76",
	deno: "1.24",
	edge: "79",
	electron: "6.0",
	firefox: "71",
	"firefox-android": "71",
	ios: "13.0",
	node: "12.9",
	oculus: "7.0",
	opera: "63",
	"opera-android": "54",
	opera_mobile: "54",
	quest: "7.0",
	rhino: "1.7.15",
	safari: "13",
	samsung: "12.0"
},
	"esnext.promise.any": {
	android: "85",
	bun: "0.1.1",
	chrome: "85",
	"chrome-android": "85",
	deno: "1.24",
	edge: "85",
	electron: "10.0",
	firefox: "79",
	"firefox-android": "79",
	ios: "14.0",
	node: "15.0",
	oculus: "12.0",
	opera: "71",
	"opera-android": "60",
	opera_mobile: "60",
	quest: "12.0",
	rhino: "1.8.0",
	safari: "14.0",
	samsung: "14.0"
},
	"esnext.promise.try": {
	android: "128",
	bun: "1.1.22",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "134",
	"firefox-android": "134",
	ios: "18.2",
	node: "23.0",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	rhino: "1.9.0",
	safari: "18.2",
	samsung: "28.0"
},
	"esnext.promise.with-resolvers": {
	android: "119",
	bun: "0.7.1",
	chrome: "119",
	"chrome-android": "119",
	deno: "1.38",
	edge: "119",
	electron: "28.0",
	firefox: "121",
	"firefox-android": "121",
	ios: "17.4",
	node: "22.0",
	oculus: "31.0",
	opera: "105",
	"opera-android": "79",
	opera_mobile: "79",
	quest: "31.0",
	safari: "17.4",
	samsung: "25.0"
},
	"esnext.reflect.define-metadata": {
},
	"esnext.reflect.delete-metadata": {
},
	"esnext.reflect.get-metadata": {
},
	"esnext.reflect.get-metadata-keys": {
},
	"esnext.reflect.get-own-metadata": {
},
	"esnext.reflect.get-own-metadata-keys": {
},
	"esnext.reflect.has-metadata": {
},
	"esnext.reflect.has-own-metadata": {
},
	"esnext.reflect.metadata": {
},
	"esnext.regexp.escape": {
	android: "136",
	bun: "1.1.22",
	chrome: "136",
	"chrome-android": "136",
	deno: "2.3.2",
	edge: "136",
	electron: "36.0",
	firefox: "134",
	"firefox-android": "134",
	ios: "18.2",
	node: "24.0",
	oculus: "39.0",
	opera: "120",
	"opera-android": "90",
	opera_mobile: "90",
	quest: "39.0",
	safari: "18.2",
	samsung: "29.0"
},
	"esnext.set.add-all": {
},
	"esnext.set.delete-all": {
},
	"esnext.set.difference.v2": {
	android: "128",
	bun: "1.2.5",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "26.0",
	node: "22.10",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	safari: "26.0",
	samsung: "28.0"
},
	"esnext.set.difference": {
},
	"esnext.set.every": {
},
	"esnext.set.filter": {
},
	"esnext.set.find": {
},
	"esnext.set.from": {
},
	"esnext.set.intersection.v2": {
	android: "128",
	bun: "1.1.1",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "18.0",
	node: "22.10",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	safari: "18.0",
	samsung: "28.0"
},
	"esnext.set.intersection": {
},
	"esnext.set.is-disjoint-from.v2": {
	android: "128",
	bun: "1.1.1",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "18.0",
	node: "22.10",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	safari: "18.0",
	samsung: "28.0"
},
	"esnext.set.is-disjoint-from": {
},
	"esnext.set.is-subset-of.v2": {
	android: "128",
	bun: "1.1.1",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "18.0",
	node: "22.10",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	safari: "18.0",
	samsung: "28.0"
},
	"esnext.set.is-subset-of": {
},
	"esnext.set.is-superset-of.v2": {
	android: "128",
	bun: "1.1.1",
	chrome: "128",
	"chrome-android": "128",
	deno: "1.46",
	edge: "128",
	electron: "32.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "18.0",
	node: "22.10",
	oculus: "35.0",
	opera: "114",
	"opera-android": "85",
	opera_mobile: "85",
	quest: "35.0",
	safari: "18.0",
	samsung: "28.0"
},
	"esnext.set.is-superset-of": {
},
	"esnext.set.join": {
},
	"esnext.set.map": {
},
	"esnext.set.of": {
},
	"esnext.set.reduce": {
},
	"esnext.set.some": {
},
	"esnext.set.symmetric-difference.v2": {
	android: "123",
	bun: "1.2.5",
	chrome: "123",
	"chrome-android": "123",
	deno: "1.41.3",
	edge: "123",
	electron: "30.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "26.0",
	node: "22.0",
	oculus: "33.0",
	opera: "109",
	"opera-android": "82",
	opera_mobile: "82",
	quest: "33.0",
	safari: "26.0",
	samsung: "27.0"
},
	"esnext.set.symmetric-difference": {
},
	"esnext.set.union.v2": {
	android: "123",
	bun: "1.2.5",
	chrome: "123",
	"chrome-android": "123",
	deno: "1.41.3",
	edge: "123",
	electron: "30.0",
	firefox: "127",
	"firefox-android": "127",
	ios: "26.0",
	node: "22.0",
	oculus: "33.0",
	opera: "109",
	"opera-android": "82",
	opera_mobile: "82",
	quest: "33.0",
	safari: "26.0",
	samsung: "27.0"
},
	"esnext.set.union": {
},
	"esnext.string.at": {
},
	"esnext.string.cooked": {
},
	"esnext.string.code-points": {
},
	"esnext.string.dedent": {
},
	"esnext.string.is-well-formed": {
	android: "111",
	bun: "0.4.0",
	chrome: "111",
	"chrome-android": "111",
	deno: "1.32",
	edge: "111",
	electron: "24.0",
	firefox: "119",
	"firefox-android": "119",
	ios: "16.4",
	node: "20.0",
	oculus: "27.0",
	opera: "97",
	"opera-android": "75",
	opera_mobile: "75",
	quest: "27.0",
	rhino: "1.8.0",
	safari: "16.4",
	samsung: "22.0"
},
	"esnext.string.match-all": {
	android: "80",
	bun: "0.1.1",
	chrome: "80",
	"chrome-android": "80",
	deno: "1.0",
	edge: "80",
	electron: "8.0",
	firefox: "73",
	"firefox-android": "73",
	hermes: "0.6",
	ios: "13.4",
	node: "14.0",
	oculus: "9.0",
	opera: "67",
	"opera-android": "57",
	opera_mobile: "57",
	quest: "9.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "13.1",
	samsung: "13.0"
},
	"esnext.string.replace-all": {
	android: "85",
	bun: "0.1.1",
	chrome: "85",
	"chrome-android": "85",
	deno: "1.2",
	edge: "85",
	electron: "10.0",
	firefox: "77",
	"firefox-android": "77",
	hermes: "0.7",
	ios: "13.4",
	node: "15.0",
	oculus: "12.0",
	opera: "71",
	"opera-android": "60",
	opera_mobile: "60",
	quest: "12.0",
	"react-native": "0.69",
	rhino: "1.7.15",
	safari: "13.1",
	samsung: "14.0"
},
	"esnext.string.to-well-formed": {
	android: "111",
	bun: "0.5.7",
	chrome: "111",
	"chrome-android": "111",
	deno: "1.32",
	edge: "111",
	electron: "24.0",
	firefox: "119",
	"firefox-android": "119",
	ios: "16.4",
	node: "20.0",
	oculus: "27.0",
	opera: "97",
	"opera-android": "75",
	opera_mobile: "75",
	quest: "27.0",
	rhino: "1.8.0",
	safari: "16.4",
	samsung: "22.0"
},
	"esnext.symbol.async-dispose": {
	android: "127",
	bun: "1.0.23",
	chrome: "127",
	"chrome-android": "127",
	deno: "1.38",
	edge: "127",
	electron: "32.0",
	firefox: "135",
	"firefox-android": "135",
	node: "20.5.0",
	oculus: "35.0",
	opera: "113",
	"opera-android": "84",
	opera_mobile: "84",
	quest: "35.0",
	samsung: "28.0"
},
	"esnext.symbol.custom-matcher": {
},
	"esnext.symbol.dispose": {
	android: "125",
	bun: "1.0.23",
	chrome: "125",
	"chrome-android": "125",
	deno: "1.38",
	edge: "125",
	electron: "31.0",
	firefox: "135",
	"firefox-android": "135",
	node: "20.5.0",
	oculus: "34.0",
	opera: "111",
	"opera-android": "83",
	opera_mobile: "83",
	quest: "34.0",
	samsung: "27.0"
},
	"esnext.symbol.is-registered-symbol": {
},
	"esnext.symbol.is-registered": {
},
	"esnext.symbol.is-well-known-symbol": {
},
	"esnext.symbol.is-well-known": {
},
	"esnext.symbol.matcher": {
},
	"esnext.symbol.metadata": {
	deno: "1.40.4"
},
	"esnext.symbol.metadata-key": {
},
	"esnext.symbol.observable": {
},
	"esnext.symbol.pattern-match": {
},
	"esnext.symbol.replace-all": {
},
	"esnext.typed-array.from-async": {
},
	"esnext.typed-array.at": {
	android: "92",
	bun: "0.1.1",
	chrome: "92",
	"chrome-android": "92",
	deno: "1.12",
	edge: "92",
	electron: "14.0",
	firefox: "90",
	"firefox-android": "90",
	hermes: "0.13",
	ios: "15.4",
	node: "16.6",
	oculus: "17.0",
	opera: "78",
	"opera-android": "65",
	opera_mobile: "65",
	quest: "17.0",
	"react-native": "0.71",
	rhino: "1.7.15",
	safari: "15.4",
	samsung: "16.0"
},
	"esnext.typed-array.filter-out": {
},
	"esnext.typed-array.filter-reject": {
},
	"esnext.typed-array.find-last": {
	android: "97",
	bun: "0.1.1",
	chrome: "97",
	"chrome-android": "97",
	deno: "1.16",
	edge: "97",
	electron: "17.0",
	firefox: "104",
	"firefox-android": "104",
	hermes: "0.11",
	ios: "15.4",
	node: "18.0",
	oculus: "20.0",
	opera: "83",
	"opera-android": "68",
	opera_mobile: "68",
	quest: "20.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "15.4",
	samsung: "18.0"
},
	"esnext.typed-array.find-last-index": {
	android: "97",
	bun: "0.1.1",
	chrome: "97",
	"chrome-android": "97",
	deno: "1.16",
	edge: "97",
	electron: "17.0",
	firefox: "104",
	"firefox-android": "104",
	hermes: "0.11",
	ios: "15.4",
	node: "18.0",
	oculus: "20.0",
	opera: "83",
	"opera-android": "68",
	opera_mobile: "68",
	quest: "20.0",
	"react-native": "0.69",
	rhino: "1.8.0",
	safari: "15.4",
	samsung: "18.0"
},
	"esnext.typed-array.group-by": {
},
	"esnext.typed-array.to-reversed": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"esnext.typed-array.to-sorted": {
	android: "110",
	bun: "0.1.1",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "16.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	rhino: "1.8.0",
	safari: "16.0",
	samsung: "21.0"
},
	"esnext.typed-array.to-spliced": {
},
	"esnext.typed-array.unique-by": {
},
	"esnext.typed-array.with": {
	android: "110",
	bun: "1.2.18",
	chrome: "110",
	"chrome-android": "110",
	deno: "1.27",
	edge: "110",
	electron: "23.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "26.0",
	node: "20.0",
	oculus: "26.0",
	opera: "96",
	"opera-android": "74",
	opera_mobile: "74",
	quest: "26.0",
	rhino: "1.8.0",
	safari: "26.0",
	samsung: "21.0"
},
	"esnext.uint8-array.from-base64": {
	android: "140",
	bun: "1.2.20",
	chrome: "140",
	"chrome-android": "140",
	deno: "2.5",
	edge: "140",
	electron: "38.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "26.2",
	node: "25.0",
	oculus: "41.0",
	opera: "124",
	"opera-android": "92",
	opera_mobile: "92",
	quest: "41.0",
	safari: "26.2"
},
	"esnext.uint8-array.from-hex": {
	android: "140",
	bun: "1.1.22",
	chrome: "140",
	"chrome-android": "140",
	deno: "2.5",
	edge: "140",
	electron: "38.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "18.2",
	node: "25.0",
	oculus: "41.0",
	opera: "124",
	"opera-android": "92",
	opera_mobile: "92",
	quest: "41.0",
	safari: "18.2"
},
	"esnext.uint8-array.set-from-base64": {
	android: "140",
	bun: "1.2.20",
	chrome: "140",
	"chrome-android": "140",
	deno: "2.5",
	edge: "140",
	electron: "38.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "26.2",
	node: "25.0",
	oculus: "41.0",
	opera: "124",
	"opera-android": "92",
	opera_mobile: "92",
	quest: "41.0",
	safari: "26.2"
},
	"esnext.uint8-array.set-from-hex": {
	android: "144",
	bun: "1.1.22",
	chrome: "144",
	"chrome-android": "144",
	edge: "144",
	electron: "40.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "18.2",
	opera: "128",
	safari: "18.2"
},
	"esnext.uint8-array.to-base64": {
	android: "140",
	bun: "1.1.22",
	chrome: "140",
	"chrome-android": "140",
	deno: "2.5",
	edge: "140",
	electron: "38.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "18.2",
	node: "25.0",
	oculus: "41.0",
	opera: "124",
	"opera-android": "92",
	opera_mobile: "92",
	quest: "41.0",
	safari: "18.2"
},
	"esnext.uint8-array.to-hex": {
	android: "140",
	bun: "1.1.22",
	chrome: "140",
	"chrome-android": "140",
	deno: "2.5",
	edge: "140",
	electron: "38.0",
	firefox: "133",
	"firefox-android": "133",
	ios: "18.2",
	node: "25.0",
	oculus: "41.0",
	opera: "124",
	"opera-android": "92",
	opera_mobile: "92",
	quest: "41.0",
	safari: "18.2"
},
	"esnext.weak-map.delete-all": {
},
	"esnext.weak-map.from": {
},
	"esnext.weak-map.of": {
},
	"esnext.weak-map.emplace": {
},
	"esnext.weak-map.get-or-insert": {
	android: "145",
	bun: "1.2.20",
	chrome: "145",
	"chrome-android": "145",
	edge: "145",
	electron: "41.0",
	firefox: "144",
	"firefox-android": "144",
	ios: "26.2",
	opera: "129",
	safari: "26.2"
},
	"esnext.weak-map.get-or-insert-computed": {
	android: "145",
	bun: "1.2.20",
	chrome: "145",
	"chrome-android": "145",
	edge: "145",
	electron: "41.0",
	firefox: "144",
	"firefox-android": "144",
	ios: "26.2",
	opera: "129",
	safari: "26.2"
},
	"esnext.weak-map.upsert": {
},
	"esnext.weak-set.add-all": {
},
	"esnext.weak-set.delete-all": {
},
	"esnext.weak-set.from": {
},
	"esnext.weak-set.of": {
},
	"web.atob": {
	android: "37",
	bun: "0.1.1",
	chrome: "34",
	"chrome-android": "34",
	deno: "1.0",
	edge: "16",
	electron: "0.20",
	firefox: "27",
	"firefox-android": "27",
	hermes: "0.13",
	ios: "10.3",
	node: "18.0",
	oculus: "3.0",
	opera: "10.5",
	"opera-android": "10.5",
	opera_mobile: "10.5",
	quest: "3.0",
	"react-native": "0.74",
	safari: "10.1",
	samsung: "2.0"
},
	"web.btoa": {
	android: "3.0",
	bun: "0.1.1",
	chrome: "4",
	"chrome-android": "18",
	deno: "1.0",
	edge: "16",
	electron: "0.20",
	firefox: "27",
	"firefox-android": "27",
	ios: "1.0",
	node: "17.5",
	oculus: "3.0",
	opera: "10.5",
	"opera-android": "10.5",
	opera_mobile: "10.5",
	phantom: "1.9",
	quest: "3.0",
	safari: "3.0",
	samsung: "1.0"
},
	"web.dom-collections.for-each": {
	android: "58",
	bun: "0.1.1",
	chrome: "58",
	"chrome-android": "58",
	deno: "1.0",
	edge: "16",
	electron: "1.7",
	firefox: "50",
	"firefox-android": "50",
	hermes: "0.1",
	ios: "10.0",
	node: "0.0.1",
	oculus: "4.0",
	opera: "45",
	"opera-android": "43",
	opera_mobile: "43",
	quest: "4.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "10.0",
	samsung: "7.0"
},
	"web.dom-collections.iterator": {
	android: "66",
	bun: "0.1.1",
	chrome: "66",
	"chrome-android": "66",
	deno: "1.0",
	edge: "79",
	electron: "3.0",
	firefox: "60",
	"firefox-android": "60",
	hermes: "0.1",
	ios: "13.4",
	node: "0.0.1",
	oculus: "5.0",
	opera: "53",
	"opera-android": "47",
	opera_mobile: "47",
	quest: "5.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "13.1",
	samsung: "9.0"
},
	"web.dom-exception.constructor": {
	android: "46",
	bun: "0.1.1",
	chrome: "46",
	"chrome-android": "46",
	deno: "1.7",
	edge: "79",
	electron: "0.36",
	firefox: "37",
	"firefox-android": "37",
	ios: "11.3",
	node: "17.0",
	oculus: "3.0",
	opera: "33",
	"opera-android": "33",
	opera_mobile: "33",
	quest: "3.0",
	safari: "11.1",
	samsung: "5.0"
},
	"web.dom-exception.stack": {
	deno: "1.15",
	firefox: "37",
	"firefox-android": "37",
	node: "17.0"
},
	"web.dom-exception.to-string-tag": {
	android: "49",
	bun: "0.1.1",
	chrome: "49",
	"chrome-android": "49",
	deno: "1.7",
	edge: "79",
	electron: "0.37",
	firefox: "51",
	"firefox-android": "51",
	ios: "11.3",
	node: "17.0",
	oculus: "3.0",
	opera: "36",
	"opera-android": "36",
	opera_mobile: "36",
	quest: "3.0",
	safari: "11.1",
	samsung: "5.0"
},
	"web.immediate": {
	bun: "0.4.0",
	deno: "2.4",
	ie: "10",
	node: "0.9.1"
},
	"web.queue-microtask": {
	android: "71",
	chrome: "71",
	"chrome-android": "71",
	deno: "1.0",
	edge: "79",
	electron: "5.0",
	firefox: "69",
	"firefox-android": "69",
	ios: "12.2",
	node: "12.0",
	oculus: "6.0",
	opera: "58",
	"opera-android": "50",
	opera_mobile: "50",
	quest: "6.0",
	safari: "12.1",
	samsung: "10.0"
},
	"web.self": {
	android: "86",
	bun: "1.0.22",
	chrome: "86",
	"chrome-android": "86",
	deno: "1.46.0",
	edge: "86",
	electron: "11.0",
	firefox: "31",
	"firefox-android": "31",
	ios: "10.0",
	oculus: "12.0",
	opera: "72",
	"opera-android": "61",
	opera_mobile: "61",
	quest: "12.0",
	safari: "10",
	samsung: "14.0"
},
	"web.structured-clone": {
},
	"web.timers": {
	android: "1.5",
	bun: "0.4.0",
	chrome: "1",
	"chrome-android": "18",
	deno: "1.0",
	edge: "12",
	electron: "0.20",
	firefox: "1",
	"firefox-android": "4",
	hermes: "0.1",
	ie: "10",
	ios: "1.0",
	node: "0.0.1",
	oculus: "3.0",
	opera: "7",
	"opera-android": "7",
	opera_mobile: "7",
	phantom: "1.9",
	quest: "3.0",
	"react-native": "0.69",
	rhino: "1.7.13",
	safari: "1.0",
	samsung: "1.0"
},
	"web.url": {
	android: "67",
	bun: "0.1.1",
	chrome: "67",
	"chrome-android": "67",
	deno: "1.0",
	edge: "79",
	electron: "4.0",
	firefox: "57",
	"firefox-android": "57",
	ios: "14.0",
	node: "10.0",
	oculus: "6.0",
	opera: "54",
	"opera-android": "48",
	opera_mobile: "48",
	quest: "6.0",
	safari: "14.0",
	samsung: "9.0"
},
	"web.url.can-parse": {
	android: "120",
	bun: "1.1.0",
	chrome: "120",
	"chrome-android": "120",
	deno: "1.33.2",
	edge: "120",
	electron: "28.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "17.0",
	node: "20.1",
	oculus: "31.0",
	opera: "106",
	"opera-android": "80",
	opera_mobile: "80",
	quest: "31.0",
	safari: "17.0",
	samsung: "25.0"
},
	"web.url.parse": {
	android: "126",
	bun: "1.1.4",
	chrome: "126",
	"chrome-android": "126",
	deno: "1.43",
	edge: "126",
	electron: "31.0",
	firefox: "126",
	"firefox-android": "126",
	ios: "18.0",
	node: "22.1",
	oculus: "34.0",
	opera: "112",
	"opera-android": "83",
	opera_mobile: "83",
	quest: "34.0",
	safari: "18.0",
	samsung: "28.0"
},
	"web.url.to-json": {
	android: "71",
	bun: "0.1.1",
	chrome: "71",
	"chrome-android": "71",
	deno: "1.0",
	edge: "79",
	electron: "5.0",
	firefox: "57",
	"firefox-android": "57",
	ios: "14.0",
	node: "10.0",
	oculus: "6.0",
	opera: "58",
	"opera-android": "50",
	opera_mobile: "50",
	quest: "6.0",
	safari: "14.0",
	samsung: "10.0"
},
	"web.url-search-params": {
	android: "67",
	bun: "0.1.1",
	chrome: "67",
	"chrome-android": "67",
	deno: "1.0",
	edge: "79",
	electron: "4.0",
	firefox: "57",
	"firefox-android": "57",
	ios: "14.0",
	node: "10.0",
	oculus: "6.0",
	opera: "54",
	"opera-android": "48",
	opera_mobile: "48",
	quest: "6.0",
	safari: "14.0",
	samsung: "9.0"
},
	"web.url-search-params.delete": {
	android: "118",
	bun: "1.0.31",
	chrome: "118",
	"chrome-android": "118",
	deno: "1.35",
	edge: "118",
	electron: "27.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "17.0",
	node: "20.2",
	oculus: "30.0",
	opera: "104",
	"opera-android": "79",
	opera_mobile: "79",
	quest: "30.0",
	safari: "17.0",
	samsung: "25.0"
},
	"web.url-search-params.has": {
	android: "118",
	bun: "1.0.31",
	chrome: "118",
	"chrome-android": "118",
	deno: "1.35",
	edge: "118",
	electron: "27.0",
	firefox: "115",
	"firefox-android": "115",
	ios: "17.0",
	node: "20.2",
	oculus: "30.0",
	opera: "104",
	"opera-android": "79",
	opera_mobile: "79",
	quest: "30.0",
	safari: "17.0",
	samsung: "25.0"
},
	"web.url-search-params.size": {
	android: "113",
	bun: "1.0.2",
	chrome: "113",
	"chrome-android": "113",
	deno: "1.32",
	edge: "113",
	electron: "25.0",
	firefox: "112",
	"firefox-android": "112",
	ios: "17.0",
	node: "19.8",
	oculus: "28.0",
	opera: "99",
	"opera-android": "76",
	opera_mobile: "76",
	quest: "28.0",
	safari: "17.0",
	samsung: "23.0"
}
};

const keys = Object.keys;
const plugins = filterAvailable(originalPlugins);
const pluginsBugfixes = filterAvailable(originalPluginsBugfixes);
const overlappingPlugins = filterAvailable(originalOverlappingPlugins);
function filterAvailable(data) {
  const result = {};
  for (const plugin of keys(data)) {
    if (Object.hasOwn(availablePlugins, plugin)) {
      result[plugin] = data[plugin];
    }
  }
  return result;
}

const TopLevelOptions = {
  configPath: "configPath",
  corejs: "corejs",
  debug: "debug",
  exclude: "exclude",
  forceAllTransforms: "forceAllTransforms",
  ignoreBrowserslistConfig: "ignoreBrowserslistConfig",
  include: "include",
  modules: "modules",
  shippedProposals: "shippedProposals",
  targets: "targets",
  useBuiltIns: "useBuiltIns",
  browserslistEnv: "browserslistEnv"
};
const ModulesOption = {
  false: false,
  auto: "auto",
  amd: "amd",
  commonjs: "commonjs",
  cjs: "cjs",
  systemjs: "systemjs",
  umd: "umd"
};
const UseBuiltInsOption = {
  false: false,
  entry: "entry",
  usage: "usage"
};

const v = new OptionValidator("@babel/preset-env");
const allPluginsList = [...Object.keys(plugins), ...Object.keys(pluginsBugfixes)];
const modulePlugins = ["transform-dynamic-import", ...Object.keys(moduleTransformations).map(m => moduleTransformations[m])];
const getValidIncludesAndExcludes = (type, corejs) => {
  const set = new Set(allPluginsList);
  if (type === "exclude") modulePlugins.map(set.add, set);
  if (corejs) {
    Object.keys(corejs3Polyfills).map(set.add, set);
  }
  return Array.from(set);
};
const normalizePluginName = plugin => plugin.replace(/^(?:@babel\/|babel-)(?:plugin-)?/, "");
const expandIncludesAndExcludes = (filterList = [], type, corejs) => {
  if (filterList.length === 0) return [];
  const filterableItems = getValidIncludesAndExcludes(type, corejs);
  const invalidFilters = [];
  const selectedPlugins = filterList.flatMap(filter => {
    let re;
    if (typeof filter === "string") {
      try {
        re = new RegExp(`^${normalizePluginName(filter)}$`);
      } catch (_) {
        invalidFilters.push(filter);
        return [];
      }
    } else {
      re = filter;
    }
    const items = filterableItems.filter(item => {
      return re.test(item);
    });
    if (items.length === 0) invalidFilters.push(filter);
    return items;
  });
  v.invariant(invalidFilters.length === 0, `The plugins/built-ins '${invalidFilters.join(", ")}' passed to the '${type}' option are not
    valid. Please check data/[plugin-features|built-in-features].js in babel-preset-env`);
  return selectedPlugins;
};
const checkDuplicateIncludeExcludes = (include = [], exclude = []) => {
  const duplicates = include.filter(opt => exclude.includes(opt));
  v.invariant(duplicates.length === 0, `The plugins/built-ins '${duplicates.join(", ")}' were found in both the "include" and
    "exclude" options.`);
};
const normalizeTargets = targets => {
  if (typeof targets === "string" || Array.isArray(targets)) {
    return {
      browsers: targets
    };
  }
  return {
    ...targets
  };
};
const validateModulesOption = (modulesOpt = ModulesOption.auto) => {
  v.invariant(ModulesOption[modulesOpt.toString()] || modulesOpt === ModulesOption.false, `The 'modules' option must be one of \n` + ` - 'false' to indicate no module processing\n` + ` - a specific module type: 'commonjs', 'amd', 'umd', 'systemjs'` + ` - 'auto' (default) which will automatically select 'commonjs' if the current\n` + `   process is known to *not* support ES module syntax, or 'false' otherwise\n`);
  return modulesOpt;
};
const validateUseBuiltInsOption = (builtInsOpt = false) => {
  v.invariant(UseBuiltInsOption[builtInsOpt.toString()] || builtInsOpt === UseBuiltInsOption.false, `The 'useBuiltIns' option must be either
    'false' (default) to indicate no polyfill,
    '"entry"' to indicate replacing the entry polyfill, or
    '"usage"' to import only used polyfills per file`);
  return builtInsOpt;
};
function normalizeCoreJSOption(corejs, useBuiltIns) {
  let proposals = false;
  let rawVersion;
  {
    rawVersion = corejs;
  }
  const version = rawVersion ? semver.coerce(String(rawVersion)) : false;
  if (version) {
    if (useBuiltIns) {
      if (version.major !== 3) {
        throw new RangeError("Invalid Option: The version passed to `corejs` is invalid. Currently, " + "only core-js@3 is supported.");
      }
      if (typeof rawVersion !== "string" || !String(rawVersion).includes(".")) {
        throw new Error('Invalid Option: The version passed to `corejs` is invalid. Please use string and specify the minor version, such as `"3.33"`.');
      }
    } else {
      console.warn("\nWARNING (@babel/preset-env): The `corejs` option only has an effect when the `useBuiltIns` option is not `false`\n");
    }
  }
  return {
    version,
    proposals
  };
}
function normalizeOptions(opts) {
  v.invariant(!Object.hasOwn(opts, "bugfixes"), "The 'bugfixes' option has been removed, and now bugfix plugins are" + " always enabled. Please remove it from your config.");
  v.validateTopLevelOptions(opts, TopLevelOptions);
  if (opts.useBuiltIns) {
    throw new Error("The 'useBuiltIns' option has been removed. Please use babel-plugin-polyfill-corejs3 instead.");
  }
  const useBuiltIns = validateUseBuiltInsOption(false);
  const corejs = normalizeCoreJSOption(null, useBuiltIns);
  const include = expandIncludesAndExcludes(opts.include, TopLevelOptions.include, !!corejs.version && corejs.version.major);
  const exclude = expandIncludesAndExcludes(opts.exclude, TopLevelOptions.exclude, !!corejs.version && corejs.version.major);
  checkDuplicateIncludeExcludes(include, exclude);
  return {
    configPath: v.validateStringOption(TopLevelOptions.configPath, opts.configPath, process.cwd()),
    corejs,
    debug: v.validateBooleanOption(TopLevelOptions.debug, opts.debug, false),
    include,
    exclude,
    forceAllTransforms: v.validateBooleanOption(TopLevelOptions.forceAllTransforms, opts.forceAllTransforms, false),
    ignoreBrowserslistConfig: v.validateBooleanOption(TopLevelOptions.ignoreBrowserslistConfig, opts.ignoreBrowserslistConfig, false),
    modules: validateModulesOption(opts.modules),
    shippedProposals: v.validateBooleanOption(TopLevelOptions.shippedProposals, opts.shippedProposals, false),
    targets: normalizeTargets(opts.targets),
    useBuiltIns: useBuiltIns,
    browserslistEnv: v.validateStringOption(TopLevelOptions.browserslistEnv, opts.browserslistEnv)
  };
}

const proposalPlugins = new Set([]);
const proposalSyntaxPlugins = [];
const pluginSyntaxObject = {};
const pluginSyntaxEntries = Object.keys(pluginSyntaxObject).map(function (key) {
  return [key, pluginSyntaxObject[key]];
});
const pluginSyntaxMap = new Map(pluginSyntaxEntries);

function isPluginRequired(targets, support) {
  throw new Error("`isPluginRequired` has been removed in Babel 8. Please use `isRequired` from `@babel/helper-compilation-targets` instead.");
}
function filterStageFromList(list, stageList) {
  return Object.keys(list).reduce((result, item) => {
    if (!stageList.has(item)) {
      result[item] = list[item];
    }
    return result;
  }, {});
}
const pluginsListWithProposals = Object.assign({}, plugins, pluginsBugfixes);
const pluginsListWithoutProposals = filterStageFromList(pluginsListWithProposals, proposalPlugins);
const getPlugin = pluginName => {
  const plugin = availablePlugins[pluginName]();
  if (!plugin) {
    throw new Error(`Could not find plugin "${pluginName}". Ensure there is an entry in ./available-plugins.js for it.`);
  }
  return plugin;
};
const transformIncludesAndExcludes = opts => {
  return opts.reduce((result, opt) => {
    const target = /^(?:es|es6|es7|esnext|web)\./.test(opt) ? "builtIns" : "plugins";
    result[target].add(opt);
    return result;
  }, {
    all: opts,
    plugins: new Set(),
    builtIns: new Set()
  });
};
function getSpecialModulesPluginNames(modules, shouldTransformDynamicImport) {
  const modulesPluginNames = [];
  if (modules) {
    modulesPluginNames.push(moduleTransformations[modules]);
  }
  if (shouldTransformDynamicImport) {
    if (modules && modules !== "umd") {
      modulesPluginNames.push("transform-dynamic-import");
    } else {
      console.warn("Dynamic import can only be transformed when transforming ES" + " modules to AMD, CommonJS or SystemJS.");
    }
  }
  return modulesPluginNames;
}
const getCoreJSOptions = ({
  useBuiltIns,
  corejs,
  polyfillTargets,
  include,
  exclude,
  proposals,
  shippedProposals,
  debug
}) => ({
  method: `${useBuiltIns}-global`,
  version: corejs ? corejs.toString() : undefined,
  targets: polyfillTargets,
  include,
  exclude,
  proposals,
  shippedProposals,
  debug,
  "#__secret_key__@babel/preset-env__compatibility": {
    noRuntimeName: true
  }
});
function getLocalTargets(optionsTargets, ignoreBrowserslistConfig, configPath, browserslistEnv, api) {
  if (optionsTargets?.esmodules && optionsTargets.browsers) {
    console.warn(`
@babel/preset-env: esmodules and browsers targets have been specified together.
\`browsers\` target, \`${optionsTargets.browsers.toString()}\` will be ignored.
`);
  }
  return getTargets(optionsTargets, {
    ignoreBrowserslistConfig,
    configPath,
    browserslistEnv,
    onBrowserslistConfigFound(config) {
      api.addExternalDependency(config);
    }
  });
}
function supportsStaticESM(caller) {
  return caller?.supportsStaticESM ?? true;
}
function supportsDynamicImport(caller) {
  return caller?.supportsDynamicImport ?? true;
}
function supportsExportNamespaceFrom(caller) {
  return caller?.supportsExportNamespaceFrom ?? null;
}
const index = declarePreset((api, opts) => {
  api.assertVersion("^7.0.0-0 || ^8.0.0");
  const babelTargets = api.targets();
  if ("loose" in opts || "spec" in opts) {
    throw new Error("@babel/preset-env: The 'loose' and 'spec' options have been removed, " + "and you should configure granular compiler assumptions instead. See " + "https://babeljs.io/assumptions for more information.");
  }
  const {
    configPath,
    debug,
    exclude: optionsExclude,
    forceAllTransforms,
    ignoreBrowserslistConfig,
    include: optionsInclude,
    modules: optionsModules,
    shippedProposals,
    targets: optionsTargets,
    useBuiltIns,
    corejs: {
      version: corejs,
      proposals
    },
    browserslistEnv
  } = normalizeOptions(opts);
  let targets = babelTargets;
  if (semver.lt(api.version, "7.13.0") || opts.targets || opts.configPath || opts.browserslistEnv || opts.ignoreBrowserslistConfig) {
    targets = getLocalTargets(optionsTargets, ignoreBrowserslistConfig, configPath, browserslistEnv, api);
  }
  const transformTargets = forceAllTransforms ? {} : targets;
  const include = transformIncludesAndExcludes(optionsInclude);
  const exclude = transformIncludesAndExcludes(optionsExclude);
  const compatData = shippedProposals ? pluginsListWithProposals : pluginsListWithoutProposals;
  const modules = optionsModules === "auto" ? api.caller(supportsStaticESM) ? false : "commonjs" : optionsModules;
  const shouldTransformDynamicImport = optionsModules === "auto" ? !api.caller(supportsDynamicImport) : !!modules;
  if (!exclude.plugins.has("transform-export-namespace-from") && (modules !== false || optionsModules === "auto" && !api.caller(supportsExportNamespaceFrom))) {
    include.plugins.add("transform-export-namespace-from");
  }
  const pluginNames = filterItems(compatData, include.plugins, exclude.plugins, transformTargets, getSpecialModulesPluginNames(modules, shouldTransformDynamicImport), undefined, pluginSyntaxMap);
  if (shippedProposals) {
    addProposalSyntaxPlugins(pluginNames, proposalSyntaxPlugins);
  }
  removeUnsupportedItems(pluginNames, api.version);
  removeUnnecessaryItems(pluginNames, overlappingPlugins);
  const polyfillPlugins = useBuiltIns ? [[pluginCoreJS3, getCoreJSOptions({
    useBuiltIns,
    corejs,
    polyfillTargets: targets,
    include: include.builtIns,
    exclude: exclude.builtIns,
    proposals,
    shippedProposals,
    debug
  })]] : [];
  const pluginUseBuiltIns = useBuiltIns !== false;
  const plugins = Array.from(pluginNames).map(pluginName => {
    return [getPlugin(pluginName), {
      useBuiltIns: pluginUseBuiltIns
    }];
  }).concat(polyfillPlugins);
  if (debug) {
    console.log("@babel/preset-env: `DEBUG` option");
    console.log("\nUsing targets:");
    console.log(JSON.stringify(prettifyTargets(targets), null, 2));
    console.log(`\nUsing modules transform: ${optionsModules.toString()}`);
    console.log("\nUsing plugins:");
    pluginNames.forEach(pluginName => {
      logPlugin(pluginName, targets, compatData);
    });
    if (!useBuiltIns) {
      console.log("\nUsing polyfills: No polyfills were added, since the `useBuiltIns` option was not set.");
    }
  }
  return {
    plugins
  };
});

export { index as default, isPluginRequired, transformIncludesAndExcludes };
